% Generated by Sphinx.
\def\sphinxdocclass{report}
\documentclass[letterpaper,10pt,english]{sphinxmanual}
\usepackage[utf8]{inputenc}
\DeclareUnicodeCharacter{00A0}{\nobreakspace}
\usepackage{cmap}
\usepackage[T1]{fontenc}
\usepackage{babel}
\usepackage{times}
\usepackage[Bjarne]{fncychap}
\usepackage{longtable}
\usepackage{sphinx}
\usepackage{multirow}


\title{fSeq Documentation}
\date{June 24, 2014}
\release{1.0.0a}
\author{Martin Zackrisson}
\newcommand{\sphinxlogo}{}
\renewcommand{\releasename}{Release}
\makeindex

\makeatletter
\def\PYG@reset{\let\PYG@it=\relax \let\PYG@bf=\relax%
    \let\PYG@ul=\relax \let\PYG@tc=\relax%
    \let\PYG@bc=\relax \let\PYG@ff=\relax}
\def\PYG@tok#1{\csname PYG@tok@#1\endcsname}
\def\PYG@toks#1+{\ifx\relax#1\empty\else%
    \PYG@tok{#1}\expandafter\PYG@toks\fi}
\def\PYG@do#1{\PYG@bc{\PYG@tc{\PYG@ul{%
    \PYG@it{\PYG@bf{\PYG@ff{#1}}}}}}}
\def\PYG#1#2{\PYG@reset\PYG@toks#1+\relax+\PYG@do{#2}}

\expandafter\def\csname PYG@tok@gd\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.63,0.00,0.00}{##1}}}
\expandafter\def\csname PYG@tok@gu\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.50,0.00,0.50}{##1}}}
\expandafter\def\csname PYG@tok@gt\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.00,0.27,0.87}{##1}}}
\expandafter\def\csname PYG@tok@gs\endcsname{\let\PYG@bf=\textbf}
\expandafter\def\csname PYG@tok@gr\endcsname{\def\PYG@tc##1{\textcolor[rgb]{1.00,0.00,0.00}{##1}}}
\expandafter\def\csname PYG@tok@cm\endcsname{\let\PYG@it=\textit\def\PYG@tc##1{\textcolor[rgb]{0.25,0.50,0.56}{##1}}}
\expandafter\def\csname PYG@tok@vg\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.73,0.38,0.84}{##1}}}
\expandafter\def\csname PYG@tok@m\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.13,0.50,0.31}{##1}}}
\expandafter\def\csname PYG@tok@mh\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.13,0.50,0.31}{##1}}}
\expandafter\def\csname PYG@tok@cs\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.25,0.50,0.56}{##1}}\def\PYG@bc##1{\setlength{\fboxsep}{0pt}\colorbox[rgb]{1.00,0.94,0.94}{\strut ##1}}}
\expandafter\def\csname PYG@tok@ge\endcsname{\let\PYG@it=\textit}
\expandafter\def\csname PYG@tok@vc\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.73,0.38,0.84}{##1}}}
\expandafter\def\csname PYG@tok@il\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.13,0.50,0.31}{##1}}}
\expandafter\def\csname PYG@tok@go\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.20,0.20,0.20}{##1}}}
\expandafter\def\csname PYG@tok@cp\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@gi\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.00,0.63,0.00}{##1}}}
\expandafter\def\csname PYG@tok@gh\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.00,0.00,0.50}{##1}}}
\expandafter\def\csname PYG@tok@ni\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.84,0.33,0.22}{##1}}}
\expandafter\def\csname PYG@tok@nl\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.00,0.13,0.44}{##1}}}
\expandafter\def\csname PYG@tok@nn\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.05,0.52,0.71}{##1}}}
\expandafter\def\csname PYG@tok@no\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.38,0.68,0.84}{##1}}}
\expandafter\def\csname PYG@tok@na\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.25,0.44,0.63}{##1}}}
\expandafter\def\csname PYG@tok@nb\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@nc\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.05,0.52,0.71}{##1}}}
\expandafter\def\csname PYG@tok@nd\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.33,0.33,0.33}{##1}}}
\expandafter\def\csname PYG@tok@ne\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@nf\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.02,0.16,0.49}{##1}}}
\expandafter\def\csname PYG@tok@si\endcsname{\let\PYG@it=\textit\def\PYG@tc##1{\textcolor[rgb]{0.44,0.63,0.82}{##1}}}
\expandafter\def\csname PYG@tok@s2\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.25,0.44,0.63}{##1}}}
\expandafter\def\csname PYG@tok@vi\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.73,0.38,0.84}{##1}}}
\expandafter\def\csname PYG@tok@nt\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.02,0.16,0.45}{##1}}}
\expandafter\def\csname PYG@tok@nv\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.73,0.38,0.84}{##1}}}
\expandafter\def\csname PYG@tok@s1\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.25,0.44,0.63}{##1}}}
\expandafter\def\csname PYG@tok@gp\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.78,0.36,0.04}{##1}}}
\expandafter\def\csname PYG@tok@sh\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.25,0.44,0.63}{##1}}}
\expandafter\def\csname PYG@tok@ow\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@sx\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.78,0.36,0.04}{##1}}}
\expandafter\def\csname PYG@tok@bp\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@c1\endcsname{\let\PYG@it=\textit\def\PYG@tc##1{\textcolor[rgb]{0.25,0.50,0.56}{##1}}}
\expandafter\def\csname PYG@tok@kc\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@c\endcsname{\let\PYG@it=\textit\def\PYG@tc##1{\textcolor[rgb]{0.25,0.50,0.56}{##1}}}
\expandafter\def\csname PYG@tok@mf\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.13,0.50,0.31}{##1}}}
\expandafter\def\csname PYG@tok@err\endcsname{\def\PYG@bc##1{\setlength{\fboxsep}{0pt}\fcolorbox[rgb]{1.00,0.00,0.00}{1,1,1}{\strut ##1}}}
\expandafter\def\csname PYG@tok@kd\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@ss\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.32,0.47,0.09}{##1}}}
\expandafter\def\csname PYG@tok@sr\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.14,0.33,0.53}{##1}}}
\expandafter\def\csname PYG@tok@mo\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.13,0.50,0.31}{##1}}}
\expandafter\def\csname PYG@tok@mi\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.13,0.50,0.31}{##1}}}
\expandafter\def\csname PYG@tok@kn\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@o\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.40,0.40,0.40}{##1}}}
\expandafter\def\csname PYG@tok@kr\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@s\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.25,0.44,0.63}{##1}}}
\expandafter\def\csname PYG@tok@kp\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@w\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.73,0.73,0.73}{##1}}}
\expandafter\def\csname PYG@tok@kt\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.56,0.13,0.00}{##1}}}
\expandafter\def\csname PYG@tok@sc\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.25,0.44,0.63}{##1}}}
\expandafter\def\csname PYG@tok@sb\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.25,0.44,0.63}{##1}}}
\expandafter\def\csname PYG@tok@k\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@se\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.25,0.44,0.63}{##1}}}
\expandafter\def\csname PYG@tok@sd\endcsname{\let\PYG@it=\textit\def\PYG@tc##1{\textcolor[rgb]{0.25,0.44,0.63}{##1}}}

\def\PYGZbs{\char`\\}
\def\PYGZus{\char`\_}
\def\PYGZob{\char`\{}
\def\PYGZcb{\char`\}}
\def\PYGZca{\char`\^}
\def\PYGZam{\char`\&}
\def\PYGZlt{\char`\<}
\def\PYGZgt{\char`\>}
\def\PYGZsh{\char`\#}
\def\PYGZpc{\char`\%}
\def\PYGZdl{\char`\$}
\def\PYGZhy{\char`\-}
\def\PYGZsq{\char`\'}
\def\PYGZdq{\char`\"}
\def\PYGZti{\char`\~}
% for compatibility with earlier versions
\def\PYGZat{@}
\def\PYGZlb{[}
\def\PYGZrb{]}
\makeatother

\begin{document}

\maketitle
\tableofcontents
\phantomsection\label{index::doc}


fSeq is an extensible toolkit aimed at fast reading of sequence data with a
common interface for all types of sequences.
Upon that, the ability to encode the contents of the data into suitable numeric
formats prepares for the second part.

After reading contents \emph{fSeq} allows for post-processing and report-making as
integral parts of its framework.
\emph{fSeq} comes with a small set of report builders and report makers, but these
have equally been designed to allow fast development of further capabilities.


\chapter{Contents}
\label{index:welcome-to-fseq}\label{index:contents}

\section{fseq}
\label{modules::doc}\label{modules:fseq}

\subsection{fseq package}
\label{fseq:fseq-package}\label{fseq::doc}

\subsubsection{Subpackages}
\label{fseq:subpackages}

\paragraph{fseq.reading package}
\label{fseq.reading:fseq-reading-package}\label{fseq.reading::doc}

\subparagraph{Submodules}
\label{fseq.reading:submodules}

\subparagraph{fseq.reading.seq\_encoder module}
\label{fseq.reading:module-fseq.reading.seq_encoder}\label{fseq.reading:fseq-reading-seq-encoder-module}\index{fseq.reading.seq\_encoder (module)}\index{FastQ (class in fseq.reading.seq\_encoder)}

\begin{fulllineitems}
\phantomsection\label{fseq.reading:fseq.reading.seq_encoder.FastQ}\pysigline{\strong{class }\code{fseq.reading.seq\_encoder.}\bfcode{FastQ}}
Bases: {\hyperref[fseq.reading:fseq.reading.seq_encoder.SeqFormat]{\code{fseq.reading.seq\_encoder.SeqFormat}}}

Detector of FASTQ format
\begin{description}
\item[{\textbf{Note:} This class can be subclassed to detect the different}] \leavevmode
quality-encoding schemes in the future.

\end{description}


\strong{See also:}

\begin{description}
\item[{{\hyperref[fseq.reading:fseq.reading.seq_encoder.SeqFormat]{\code{SeqFormat}}}}] \leavevmode
Base class for detectors

\end{description}


\paragraph{Examples}

Valid format:

\begin{Verbatim}[commandchars=\\\{\}]
@Contig\PYGZus{}1
AACAATACGA
+Contig\PYGZus{}1
@+\PYGZgt{}AACADGH
@Contig\PYGZus{}2
CCATTTACGA
+
\PYGZgt{}CCAGJDFGH
\end{Verbatim}
\paragraph{Attributes}

\begin{tabulary}{\linewidth}{|L|L|}
\hline

HEADER\_LINE
 & \\
\hline
SEUENCE\_LINE
 & \\
\hline
QUALITY\_LINE
 & \\
\hline\end{tabulary}

\index{HEADER\_LINE (fseq.reading.seq\_encoder.FastQ attribute)}

\begin{fulllineitems}
\phantomsection\label{fseq.reading:fseq.reading.seq_encoder.FastQ.HEADER_LINE}\pysigline{\bfcode{HEADER\_LINE}\strong{ = 0}}
\end{fulllineitems}

\index{QUALITY\_LINE (fseq.reading.seq\_encoder.FastQ attribute)}

\begin{fulllineitems}
\phantomsection\label{fseq.reading:fseq.reading.seq_encoder.FastQ.QUALITY_LINE}\pysigline{\bfcode{QUALITY\_LINE}\strong{ = 3}}
\end{fulllineitems}

\index{SEQUENCE\_LINE (fseq.reading.seq\_encoder.FastQ attribute)}

\begin{fulllineitems}
\phantomsection\label{fseq.reading:fseq.reading.seq_encoder.FastQ.SEQUENCE_LINE}\pysigline{\bfcode{SEQUENCE\_LINE}\strong{ = 1}}
\end{fulllineitems}

\index{expects() (fseq.reading.seq\_encoder.FastQ method)}

\begin{fulllineitems}
\phantomsection\label{fseq.reading:fseq.reading.seq_encoder.FastQ.expects}\pysiglinewithargsret{\bfcode{expects}}{\emph{line}}{}
Test for if line fits into required pattern for the format
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\textbf{line: str}
\begin{quote}

A line as read from file
\end{quote}

\item[{Returns}] \leavevmode
bool

\item[{Raises}] \leavevmode
\textbf{FormatImplementationError}
\begin{quote}

If \code{SeqFormat.expects} (the base class method) is called.
\end{quote}

\end{description}\end{quote}

\end{fulllineitems}

\index{hasQuality (fseq.reading.seq\_encoder.FastQ attribute)}

\begin{fulllineitems}
\phantomsection\label{fseq.reading:fseq.reading.seq_encoder.FastQ.hasQuality}\pysigline{\bfcode{hasQuality}}
If format quality information: bool

\end{fulllineitems}

\index{hasSequence (fseq.reading.seq\_encoder.FastQ attribute)}

\begin{fulllineitems}
\phantomsection\label{fseq.reading:fseq.reading.seq_encoder.FastQ.hasSequence}\pysigline{\bfcode{hasSequence}}
If format has sequence information: bool

\end{fulllineitems}

\index{itemSize (fseq.reading.seq\_encoder.FastQ attribute)}

\begin{fulllineitems}
\phantomsection\label{fseq.reading:fseq.reading.seq_encoder.FastQ.itemSize}\pysigline{\bfcode{itemSize}}
The size (lines) of each entry: int

\end{fulllineitems}

\index{name (fseq.reading.seq\_encoder.FastQ attribute)}

\begin{fulllineitems}
\phantomsection\label{fseq.reading:fseq.reading.seq_encoder.FastQ.name}\pysigline{\bfcode{name}}
Human readable description of format: str

\end{fulllineitems}


\end{fulllineitems}

\index{FastaMultiline (class in fseq.reading.seq\_encoder)}

\begin{fulllineitems}
\phantomsection\label{fseq.reading:fseq.reading.seq_encoder.FastaMultiline}\pysigline{\strong{class }\code{fseq.reading.seq\_encoder.}\bfcode{FastaMultiline}}
Bases: {\hyperref[fseq.reading:fseq.reading.seq_encoder.SeqFormat]{\code{fseq.reading.seq\_encoder.SeqFormat}}}

Detector of multi-line FASTA
\begin{description}
\item[{\textbf{Note:} This format is not supported in the current implementation}] \leavevmode
as it has no predictable item-size.

\end{description}


\strong{See also:}

\begin{description}
\item[{{\hyperref[fseq.reading:fseq.reading.seq_encoder.SeqFormat]{\code{SeqFormat}}}}] \leavevmode
Base class for detectors

\item[{{\hyperref[fseq.reading:fseq.reading.seq_encoder.FastaSingleline]{\code{FastaSingleline}}}}] \leavevmode
Derived class, with more restricitons

\end{description}


\paragraph{Examples}

Valid format:

\begin{Verbatim}[commandchars=\\\{\}]
\PYGZgt{}Contig\PYGZus{}1
ACAATACA
GATTACA
\PYGZgt{}Contig\PYGZus{}2
ACCCACA
\PYGZgt{}Contig\PYGZus{}3
ACCAAACA
CCAACACA
ACAACCAC
\end{Verbatim}
\paragraph{Attributes}

\begin{tabulary}{\linewidth}{|L|L|}
\hline

HEADER\_LINE
 & \\
\hline
SEUENCE\_LINE
 & \\
\hline
QUALITY\_LINE
 & \\
\hline\end{tabulary}

\index{HEADER\_LINE (fseq.reading.seq\_encoder.FastaMultiline attribute)}

\begin{fulllineitems}
\phantomsection\label{fseq.reading:fseq.reading.seq_encoder.FastaMultiline.HEADER_LINE}\pysigline{\bfcode{HEADER\_LINE}\strong{ = 0}}
\end{fulllineitems}

\index{QUALITY\_LINE (fseq.reading.seq\_encoder.FastaMultiline attribute)}

\begin{fulllineitems}
\phantomsection\label{fseq.reading:fseq.reading.seq_encoder.FastaMultiline.QUALITY_LINE}\pysigline{\bfcode{QUALITY\_LINE}\strong{ = None}}
\end{fulllineitems}

\index{SEQUENCE\_LINE (fseq.reading.seq\_encoder.FastaMultiline attribute)}

\begin{fulllineitems}
\phantomsection\label{fseq.reading:fseq.reading.seq_encoder.FastaMultiline.SEQUENCE_LINE}\pysigline{\bfcode{SEQUENCE\_LINE}\strong{ = -1}}
\end{fulllineitems}

\index{expects() (fseq.reading.seq\_encoder.FastaMultiline method)}

\begin{fulllineitems}
\phantomsection\label{fseq.reading:fseq.reading.seq_encoder.FastaMultiline.expects}\pysiglinewithargsret{\bfcode{expects}}{\emph{line}}{}
Test for if line fits into required pattern for the format
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\textbf{line: str}
\begin{quote}

A line as read from file
\end{quote}

\item[{Returns}] \leavevmode
bool

\item[{Raises}] \leavevmode
\textbf{FormatImplementationError}
\begin{quote}

If \code{SeqFormat.expects} (the base class method) is called.
\end{quote}

\end{description}\end{quote}

\end{fulllineitems}

\index{hasQuality (fseq.reading.seq\_encoder.FastaMultiline attribute)}

\begin{fulllineitems}
\phantomsection\label{fseq.reading:fseq.reading.seq_encoder.FastaMultiline.hasQuality}\pysigline{\bfcode{hasQuality}}
If format quality information: bool

\end{fulllineitems}

\index{hasSequence (fseq.reading.seq\_encoder.FastaMultiline attribute)}

\begin{fulllineitems}
\phantomsection\label{fseq.reading:fseq.reading.seq_encoder.FastaMultiline.hasSequence}\pysigline{\bfcode{hasSequence}}
If format has sequence information: bool

\end{fulllineitems}

\index{name (fseq.reading.seq\_encoder.FastaMultiline attribute)}

\begin{fulllineitems}
\phantomsection\label{fseq.reading:fseq.reading.seq_encoder.FastaMultiline.name}\pysigline{\bfcode{name}}
Human readable description of format: str

\end{fulllineitems}


\end{fulllineitems}

\index{FastaSingleline (class in fseq.reading.seq\_encoder)}

\begin{fulllineitems}
\phantomsection\label{fseq.reading:fseq.reading.seq_encoder.FastaSingleline}\pysigline{\strong{class }\code{fseq.reading.seq\_encoder.}\bfcode{FastaSingleline}}
Bases: {\hyperref[fseq.reading:fseq.reading.seq_encoder.FastaMultiline]{\code{fseq.reading.seq\_encoder.FastaMultiline}}}

Detector of single-line FASTA format.


\strong{See also:}

\begin{description}
\item[{{\hyperref[fseq.reading:fseq.reading.seq_encoder.SeqFormat]{\code{SeqFormat}}}}] \leavevmode
Base class for detectors

\item[{{\hyperref[fseq.reading:fseq.reading.seq_encoder.FastaMultiline]{\code{FastaMultiline}}}}] \leavevmode
Parent class

\end{description}


\paragraph{Examples}

Valid format:

\begin{Verbatim}[commandchars=\\\{\}]
\PYGZgt{}Contig\PYGZus{}1
AACAATACGA
\PYGZgt{}Contig\PYGZus{}2
CCATTTACGA
\end{Verbatim}
\paragraph{Attributes}

\begin{tabulary}{\linewidth}{|L|L|}
\hline

HEADER\_LINE
 & \\
\hline
SEUENCE\_LINE
 & \\
\hline
QUALITY\_LINE
 & \\
\hline\end{tabulary}

\index{HEADER\_LINE (fseq.reading.seq\_encoder.FastaSingleline attribute)}

\begin{fulllineitems}
\phantomsection\label{fseq.reading:fseq.reading.seq_encoder.FastaSingleline.HEADER_LINE}\pysigline{\bfcode{HEADER\_LINE}\strong{ = 0}}
\end{fulllineitems}

\index{QUALITY\_LINE (fseq.reading.seq\_encoder.FastaSingleline attribute)}

\begin{fulllineitems}
\phantomsection\label{fseq.reading:fseq.reading.seq_encoder.FastaSingleline.QUALITY_LINE}\pysigline{\bfcode{QUALITY\_LINE}\strong{ = None}}
\end{fulllineitems}

\index{SEQUENCE\_LINE (fseq.reading.seq\_encoder.FastaSingleline attribute)}

\begin{fulllineitems}
\phantomsection\label{fseq.reading:fseq.reading.seq_encoder.FastaSingleline.SEQUENCE_LINE}\pysigline{\bfcode{SEQUENCE\_LINE}\strong{ = 1}}
\end{fulllineitems}

\index{expects() (fseq.reading.seq\_encoder.FastaSingleline method)}

\begin{fulllineitems}
\phantomsection\label{fseq.reading:fseq.reading.seq_encoder.FastaSingleline.expects}\pysiglinewithargsret{\bfcode{expects}}{\emph{line}}{}
Test for if line fits into required pattern for the format
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\textbf{line: str}
\begin{quote}

A line as read from file
\end{quote}

\item[{Returns}] \leavevmode
bool

\item[{Raises}] \leavevmode
\textbf{FormatImplementationError}
\begin{quote}

If \code{SeqFormat.expects} (the base class method) is called.
\end{quote}

\end{description}\end{quote}

\end{fulllineitems}

\index{itemSize (fseq.reading.seq\_encoder.FastaSingleline attribute)}

\begin{fulllineitems}
\phantomsection\label{fseq.reading:fseq.reading.seq_encoder.FastaSingleline.itemSize}\pysigline{\bfcode{itemSize}}
The size (lines) of each entry: int

\end{fulllineitems}

\index{name (fseq.reading.seq\_encoder.FastaSingleline attribute)}

\begin{fulllineitems}
\phantomsection\label{fseq.reading:fseq.reading.seq_encoder.FastaSingleline.name}\pysigline{\bfcode{name}}
Human readable description of format: str

\end{fulllineitems}


\end{fulllineitems}

\index{FormatError}

\begin{fulllineitems}
\phantomsection\label{fseq.reading:fseq.reading.seq_encoder.FormatError}\pysigline{\strong{exception }\code{fseq.reading.seq\_encoder.}\bfcode{FormatError}}
Bases: \code{exceptions.Exception}

Sequence Format Error for exceptions relating to missmatches between
encoders and sequence formats as well as lacking formattings in encoders
and unknown sequence formats.

\end{fulllineitems}

\index{FormatImplementationError}

\begin{fulllineitems}
\phantomsection\label{fseq.reading:fseq.reading.seq_encoder.FormatImplementationError}\pysigline{\strong{exception }\code{fseq.reading.seq\_encoder.}\bfcode{FormatImplementationError}}
Bases: {\hyperref[fseq.reading:fseq.reading.seq_encoder.FormatError]{\code{fseq.reading.seq\_encoder.FormatError}}}

Error for exposing parts of interface that needs to be overwritten
in subclasses

\end{fulllineitems}

\index{FormatUnknown}

\begin{fulllineitems}
\phantomsection\label{fseq.reading:fseq.reading.seq_encoder.FormatUnknown}\pysigline{\strong{exception }\code{fseq.reading.seq\_encoder.}\bfcode{FormatUnknown}}
Bases: {\hyperref[fseq.reading:fseq.reading.seq_encoder.FormatError]{\code{fseq.reading.seq\_encoder.FormatError}}}

Error for having no available detectors left

\end{fulllineitems}

\index{SeqEncoder (class in fseq.reading.seq\_encoder)}

\begin{fulllineitems}
\phantomsection\label{fseq.reading:fseq.reading.seq_encoder.SeqEncoder}\pysiglinewithargsret{\strong{class }\code{fseq.reading.seq\_encoder.}\bfcode{SeqEncoder}}{\emph{expectedInputFormat=None}, \emph{useSequence=True}, \emph{useQuality=False}, \emph{sequenceEncoding=None}, \emph{qualityEncoding=None}, \emph{requestReports=()}}{}
Bases: \code{object}

Base class for managing encoding of raw input data.

The encoder coordinates detection of raw data format and then
extracts the information that the encoder is tasked to extract, 
convering it into suitable encoding.

\textbf{Note:} The base class is not intended to be used directly, but
to be extended.


\strong{See also:}

\begin{description}
\item[{{\hyperref[fseq.reading:fseq.reading.seq_encoder.SeqEncoderGC]{\code{SeqEncoderGC}}}}] \leavevmode
GC encoder

\end{description}


\paragraph{Attributes}

\begin{longtable}{ll}
\hline
\endfirsthead

\multicolumn{2}{c}%
{{\textsf{\tablename\ \thetable{} -- continued from previous page}}} \\
\hline
\endhead

\hline \multicolumn{2}{|r|}{{\textsf{Continued on next page}}} \\ \hline
\endfoot

\endlastfoot


{\hyperref[fseq.reading:fseq.reading.seq_encoder.SeqEncoder.format]{\code{format}}}
 & 
The input format currently expected.
\\
\hline
{\hyperref[fseq.reading:fseq.reading.seq_encoder.SeqEncoder.initiated]{\code{initiated}}}
 & 
If the sequence encoder is ready to start parsing: bool
\\
\hline
{\hyperref[fseq.reading:fseq.reading.seq_encoder.SeqEncoder.itemSize]{\code{itemSize}}}
 & 
The size in number of lines for each item in the source.
\\
\hline
{\hyperref[fseq.reading:fseq.reading.seq_encoder.SeqEncoder.qualityEncoding]{\code{qualityEncoding}}}
 & 
Map for translating quality chars to numeric values.
\\
\hline
{\hyperref[fseq.reading:fseq.reading.seq_encoder.SeqEncoder.sequenceEncoding]{\code{sequenceEncoding}}}
 & 
Map for translating sequence chars to numeric values.
\\
\hline
{\hyperref[fseq.reading:fseq.reading.seq_encoder.SeqEncoder.useQuality]{\code{useQuality}}}
 & 
If quality-line is to be used by the encoder.
\\
\hline
{\hyperref[fseq.reading:fseq.reading.seq_encoder.SeqEncoder.useSequence]{\code{useSequence}}}
 & 
If sequence-line is to be used by the encoder
\\
\hline\end{longtable}

\index{detectFormat() (fseq.reading.seq\_encoder.SeqEncoder method)}

\begin{fulllineitems}
\phantomsection\label{fseq.reading:fseq.reading.seq_encoder.SeqEncoder.detectFormat}\pysiglinewithargsret{\bfcode{detectFormat}}{}{}
Detect format based on current input stream
\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
fseq.SeqEncoder
\begin{quote}

Returns \code{self}
\end{quote}

\item[{Raises}] \leavevmode
\textbf{FormatError}
\begin{quote}

If data is not compatible with information requested
\end{quote}

\end{description}\end{quote}


\strong{See also:}

\begin{description}
\item[{{\hyperref[fseq.reading:fseq.reading.seq_encoder.SeqEncoder.detectFormat]{\code{SeqEncoder.detectFormat}}}}] \leavevmode
Guessing

\end{description}



\end{fulllineitems}

\index{feedDetection() (fseq.reading.seq\_encoder.SeqEncoder method)}

\begin{fulllineitems}
\phantomsection\label{fseq.reading:fseq.reading.seq_encoder.SeqEncoder.feedDetection}\pysiglinewithargsret{\bfcode{feedDetection}}{\emph{line}}{}
Give detection a line to work with.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\textbf{line: str}
\begin{quote}

A line from the source file
\end{quote}

\item[{Returns}] \leavevmode
fseq.SeqEncoder
\begin{quote}

Returns \code{self}
\end{quote}

\end{description}\end{quote}

\end{fulllineitems}

\index{format (fseq.reading.seq\_encoder.SeqEncoder attribute)}

\begin{fulllineitems}
\phantomsection\label{fseq.reading:fseq.reading.seq_encoder.SeqEncoder.format}\pysigline{\bfcode{format}}
The input format currently expected.
\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
SeqFormatDetector

\item[{Raises}] \leavevmode
\textbf{TypeError}
\begin{quote}

If attempting to set format with object not derived from
\code{SeqFormatDetector} or \code{SeqFormat}
\end{quote}

\end{description}\end{quote}

\end{fulllineitems}

\index{initiated (fseq.reading.seq\_encoder.SeqEncoder attribute)}

\begin{fulllineitems}
\phantomsection\label{fseq.reading:fseq.reading.seq_encoder.SeqEncoder.initiated}\pysigline{\bfcode{initiated}}
If the sequence encoder is ready to start parsing: bool

\end{fulllineitems}

\index{itemSize (fseq.reading.seq\_encoder.SeqEncoder attribute)}

\begin{fulllineitems}
\phantomsection\label{fseq.reading:fseq.reading.seq_encoder.SeqEncoder.itemSize}\pysigline{\bfcode{itemSize}}
The size in number of lines for each item in the source.
\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
int

\item[{Raises}] \leavevmode
\textbf{fseq.FormatError}
\begin{quote}

If encoder is not fully initiated
\end{quote}

\end{description}\end{quote}


\strong{See also:}

\begin{description}
\item[{{\hyperref[fseq.reading:fseq.reading.seq_encoder.SeqEncoder.initiated]{\code{SeqEncoder.initiated}}}}] \leavevmode
The initiation status

\item[{{\hyperref[fseq.reading:fseq.reading.seq_encoder.SeqEncoder.detectFormat]{\code{SeqEncoder.detectFormat}}}}] \leavevmode
Detecting the format

\item[{{\hyperref[fseq.reading:fseq.reading.seq_encoder.SeqEncoder.format]{\code{SeqEncoder.format}}}}] \leavevmode
Manually setting the format

\end{description}



\end{fulllineitems}

\index{parse() (fseq.reading.seq\_encoder.SeqEncoder method)}

\begin{fulllineitems}
\phantomsection\label{fseq.reading:fseq.reading.seq_encoder.SeqEncoder.parse}\pysiglinewithargsret{\bfcode{parse}}{\emph{lines}, \emph{out}, \emph{outindex}}{}
Placeholder parser overwritten when subclassing
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\textbf{lines: iterable of str}
\begin{quote}

Iterable of length equal to \code{self.itemSize} containing the
raw data for one item
\end{quote}

\textbf{out: numpy.ndarray}
\begin{quote}

Array that will have values written to it
\end{quote}

\textbf{outIndex: object}
\begin{quote}

Index for where the parse output should be written in the \code{out}
array such that \code{out{[}outIndex{]}} gives a sufficiently large array
that the result of parsing will fit in it.
\end{quote}

\item[{Raises}] \leavevmode
\textbf{NotImplemented}
\begin{quote}

If base class parse not overwritten or base class used directly
\end{quote}

\end{description}\end{quote}

\end{fulllineitems}

\index{qualityEncoding (fseq.reading.seq\_encoder.SeqEncoder attribute)}

\begin{fulllineitems}
\phantomsection\label{fseq.reading:fseq.reading.seq_encoder.SeqEncoder.qualityEncoding}\pysigline{\bfcode{qualityEncoding}}
Map for translating quality chars to numeric values.
\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
Object with key-lookup (implementing \code{\_\_getitem\_\_})

\item[{Raises}] \leavevmode
\textbf{TypeError}
\begin{quote}

If attempting to set with object not having char key-lookup
\end{quote}

\end{description}\end{quote}

\end{fulllineitems}

\index{requestReports (fseq.reading.seq\_encoder.SeqEncoder attribute)}

\begin{fulllineitems}
\phantomsection\label{fseq.reading:fseq.reading.seq_encoder.SeqEncoder.requestReports}\pysigline{\bfcode{requestReports}}
The reports that the encoder likes to be produced by the reader

\end{fulllineitems}

\index{reset() (fseq.reading.seq\_encoder.SeqEncoder method)}

\begin{fulllineitems}
\phantomsection\label{fseq.reading:fseq.reading.seq_encoder.SeqEncoder.reset}\pysiglinewithargsret{\bfcode{reset}}{}{}
Clears the sequence format
\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
fseq.SeqEncoder
\begin{quote}

Returns \code{self}
\end{quote}

\end{description}\end{quote}


\strong{See also:}

\begin{description}
\item[{{\hyperref[fseq.reading:fseq.reading.seq_encoder.SeqEncoder.detectFormat]{\code{SeqEncoder.detectFormat}}}}] \leavevmode
Guessing

\end{description}



\end{fulllineitems}

\index{sequenceEncoding (fseq.reading.seq\_encoder.SeqEncoder attribute)}

\begin{fulllineitems}
\phantomsection\label{fseq.reading:fseq.reading.seq_encoder.SeqEncoder.sequenceEncoding}\pysigline{\bfcode{sequenceEncoding}}
Map for translating sequence chars to numeric values.
\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
Object with key-lookup (implementing \code{\_\_getitem\_\_})

\item[{Raises}] \leavevmode
\textbf{TypeError}
\begin{quote}

If attempting to set with object not having char key-lookup
\end{quote}

\end{description}\end{quote}

\end{fulllineitems}

\index{useQuality (fseq.reading.seq\_encoder.SeqEncoder attribute)}

\begin{fulllineitems}
\phantomsection\label{fseq.reading:fseq.reading.seq_encoder.SeqEncoder.useQuality}\pysigline{\bfcode{useQuality}}
If quality-line is to be used by the encoder.
\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
bool

\end{description}\end{quote}

\end{fulllineitems}

\index{useSequence (fseq.reading.seq\_encoder.SeqEncoder attribute)}

\begin{fulllineitems}
\phantomsection\label{fseq.reading:fseq.reading.seq_encoder.SeqEncoder.useSequence}\pysigline{\bfcode{useSequence}}
If sequence-line is to be used by the encoder
\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
bool

\end{description}\end{quote}

\end{fulllineitems}


\end{fulllineitems}

\index{SeqEncoderGC (class in fseq.reading.seq\_encoder)}

\begin{fulllineitems}
\phantomsection\label{fseq.reading:fseq.reading.seq_encoder.SeqEncoderGC}\pysiglinewithargsret{\strong{class }\code{fseq.reading.seq\_encoder.}\bfcode{SeqEncoderGC}}{\emph{expectedInputFormat=None}, \emph{sequenceEncoding=None}}{}
Bases: {\hyperref[fseq.reading:fseq.reading.seq_encoder.SeqEncoder]{\code{fseq.reading.seq\_encoder.SeqEncoder}}}

GC Encoder, but useful for any sequence to numerical value encoding.

The encoder uses the sequence information of the raw data and encodes
the data according to the following:

\begin{tabulary}{\linewidth}{|L|L|}
\hline
\textsf{\relax 
Input
} & \textsf{\relax 
Encoding
}\\
\hline
G C
 & 
1.0
\\
\hline
A T
 & 
0.0
\\
\hline
N
 & 
0.5
\\
\hline\end{tabulary}


To change this behaviour, simply submit a new mappable object such as
e.g. a \code{dict} as the \code{sequenceEncoding}-parameter.
\index{parse() (fseq.reading.seq\_encoder.SeqEncoderGC method)}

\begin{fulllineitems}
\phantomsection\label{fseq.reading:fseq.reading.seq_encoder.SeqEncoderGC.parse}\pysiglinewithargsret{\bfcode{parse}}{\emph{lines}, \emph{out}, \emph{outindex}}{}
Encoder of suitable aspects of \code{lines} into \code{out}.

The sequence line of \code{lines} will be encoded onto the index
\code{outindex} of \code{out}.
If the sequence line is shorter than the data structure of
\code{out{[}outindex{]}}, the remainder of \code{out{[}outindex{]}} will be left
untouched.
If the line is longer, it will only encode up until the the length
of \code{out{[}outindex{]}}.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\textbf{lines: iterable of str}
\begin{quote}

Iterable of length equal to \code{self.itemSize} containing the
raw data for one item
\end{quote}

\textbf{out: numpy.ndarray}
\begin{quote}

Array that will have values written to it
\end{quote}

\textbf{outIndex: object}
\begin{quote}

Index for where the parse output should be written in the \code{out}
array such that \code{out{[}outIndex{]}} gives a sufficiently large array
that the result of parsing will fit in it.
\end{quote}

\end{description}\end{quote}

\end{fulllineitems}


\end{fulllineitems}

\index{SeqFormat (class in fseq.reading.seq\_encoder)}

\begin{fulllineitems}
\phantomsection\label{fseq.reading:fseq.reading.seq_encoder.SeqFormat}\pysigline{\strong{class }\code{fseq.reading.seq\_encoder.}\bfcode{SeqFormat}}
Bases: \code{object}

Base Class for implementing data format detectors.

The attributes present in subclass should overwrite the
parent properties. All subclasses must also overwrite the
\code{SeqFormat.expects(line)}.
\paragraph{Attributes}

\begin{longtable}{ll}
\hline
\endfirsthead

\multicolumn{2}{c}%
{{\textsf{\tablename\ \thetable{} -- continued from previous page}}} \\
\hline
\endhead

\hline \multicolumn{2}{|r|}{{\textsf{Continued on next page}}} \\ \hline
\endfoot

\endlastfoot


{\hyperref[fseq.reading:fseq.reading.seq_encoder.SeqFormat.name]{\code{name}}}
 & 
Human readable description of format: str
\\
\hline
{\hyperref[fseq.reading:fseq.reading.seq_encoder.SeqFormat.itemSize]{\code{itemSize}}}
 & 
The size (lines) of each entry: int
\\
\hline
{\hyperref[fseq.reading:fseq.reading.seq_encoder.SeqFormat.hasSequence]{\code{hasSequence}}}
 & 
If format has sequence information: bool
\\
\hline
{\hyperref[fseq.reading:fseq.reading.seq_encoder.SeqFormat.hasQuality]{\code{hasQuality}}}
 & 
If format quality information: bool
\\
\hline
{\hyperref[fseq.reading:fseq.reading.seq_encoder.SeqFormat.qualityEncoding]{\code{qualityEncoding}}}
 & 
If format comes with a known encoding of quality.
\\
\hline\end{longtable}


\begin{tabulary}{\linewidth}{|L|L|}
\hline

MATCH\_AA
 & \\
\hline
MATCH\_AA\_S
 & \\
\hline
MATCH\_NT
 & \\
\hline
MATCH\_NT\_S
 & \\
\hline
HEADER\_LINE
 & \\
\hline
SEUENCE\_LINE
 & \\
\hline
QUALITY\_LINE
 & \\
\hline\end{tabulary}

\index{HEADER\_LINE (fseq.reading.seq\_encoder.SeqFormat attribute)}

\begin{fulllineitems}
\phantomsection\label{fseq.reading:fseq.reading.seq_encoder.SeqFormat.HEADER_LINE}\pysigline{\bfcode{HEADER\_LINE}\strong{ = 0}}
\end{fulllineitems}

\index{MATCH\_AA (fseq.reading.seq\_encoder.SeqFormat attribute)}

\begin{fulllineitems}
\phantomsection\label{fseq.reading:fseq.reading.seq_encoder.SeqFormat.MATCH_AA}\pysigline{\bfcode{MATCH\_AA}\strong{ = \textless{}\_sre.SRE\_Pattern object at 0x2b98130a74e0\textgreater{}}}
Matches any complete line of A-Z characters allowing for asterisc at
end.

\end{fulllineitems}

\index{MATCH\_AA\_S (fseq.reading.seq\_encoder.SeqFormat attribute)}

\begin{fulllineitems}
\phantomsection\label{fseq.reading:fseq.reading.seq_encoder.SeqFormat.MATCH_AA_S}\pysigline{\bfcode{MATCH\_AA\_S}\strong{ = \textless{}\_sre.SRE\_Pattern object at 0x2b9811876ed0\textgreater{}}}
Matches as \code{MATCH\_AA\_S} but extends to include space

\end{fulllineitems}

\index{MATCH\_NT (fseq.reading.seq\_encoder.SeqFormat attribute)}

\begin{fulllineitems}
\phantomsection\label{fseq.reading:fseq.reading.seq_encoder.SeqFormat.MATCH_NT}\pysigline{\bfcode{MATCH\_NT}\strong{ = \textless{}\_sre.SRE\_Pattern object at 0x2b98130a7350\textgreater{}}}
Matches any complete line of only A T C G or N

\end{fulllineitems}

\index{MATCH\_NT\_S (fseq.reading.seq\_encoder.SeqFormat attribute)}

\begin{fulllineitems}
\phantomsection\label{fseq.reading:fseq.reading.seq_encoder.SeqFormat.MATCH_NT_S}\pysigline{\bfcode{MATCH\_NT\_S}\strong{ = \textless{}\_sre.SRE\_Pattern object at 0x2b98130a75a8\textgreater{}}}
Matches as \code{MATCH\_NT} but extends to include space

\end{fulllineitems}

\index{QUALITY\_LINE (fseq.reading.seq\_encoder.SeqFormat attribute)}

\begin{fulllineitems}
\phantomsection\label{fseq.reading:fseq.reading.seq_encoder.SeqFormat.QUALITY_LINE}\pysigline{\bfcode{QUALITY\_LINE}\strong{ = None}}
\end{fulllineitems}

\index{SEQUENCE\_LINE (fseq.reading.seq\_encoder.SeqFormat attribute)}

\begin{fulllineitems}
\phantomsection\label{fseq.reading:fseq.reading.seq_encoder.SeqFormat.SEQUENCE_LINE}\pysigline{\bfcode{SEQUENCE\_LINE}\strong{ = None}}
\end{fulllineitems}

\index{expects() (fseq.reading.seq\_encoder.SeqFormat method)}

\begin{fulllineitems}
\phantomsection\label{fseq.reading:fseq.reading.seq_encoder.SeqFormat.expects}\pysiglinewithargsret{\bfcode{expects}}{\emph{line}}{}
Test for if line fits into required pattern for the format
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\textbf{line: str}
\begin{quote}

A line as read from file
\end{quote}

\item[{Returns}] \leavevmode
bool

\item[{Raises}] \leavevmode
\textbf{FormatImplementationError}
\begin{quote}

If \code{SeqFormat.expects} (the base class method) is called.
\end{quote}

\end{description}\end{quote}

\end{fulllineitems}

\index{givenUp (fseq.reading.seq\_encoder.SeqFormat attribute)}

\begin{fulllineitems}
\phantomsection\label{fseq.reading:fseq.reading.seq_encoder.SeqFormat.givenUp}\pysigline{\bfcode{givenUp}}
Reports if format has given up even though everything still was
matching.

The purpose is to be able to promote more restricted formats that
represents a subset of the more general
\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
bool
\begin{quote}

The status
\end{quote}

\end{description}\end{quote}

\end{fulllineitems}

\index{hasQuality (fseq.reading.seq\_encoder.SeqFormat attribute)}

\begin{fulllineitems}
\phantomsection\label{fseq.reading:fseq.reading.seq_encoder.SeqFormat.hasQuality}\pysigline{\bfcode{hasQuality}}
If format quality information: bool

\end{fulllineitems}

\index{hasSequence (fseq.reading.seq\_encoder.SeqFormat attribute)}

\begin{fulllineitems}
\phantomsection\label{fseq.reading:fseq.reading.seq_encoder.SeqFormat.hasSequence}\pysigline{\bfcode{hasSequence}}
If format has sequence information: bool

\end{fulllineitems}

\index{itemSize (fseq.reading.seq\_encoder.SeqFormat attribute)}

\begin{fulllineitems}
\phantomsection\label{fseq.reading:fseq.reading.seq_encoder.SeqFormat.itemSize}\pysigline{\bfcode{itemSize}}
The size (lines) of each entry: int

\end{fulllineitems}

\index{name (fseq.reading.seq\_encoder.SeqFormat attribute)}

\begin{fulllineitems}
\phantomsection\label{fseq.reading:fseq.reading.seq_encoder.SeqFormat.name}\pysigline{\bfcode{name}}
Human readable description of format: str

\end{fulllineitems}

\index{qualityEncoding (fseq.reading.seq\_encoder.SeqFormat attribute)}

\begin{fulllineitems}
\phantomsection\label{fseq.reading:fseq.reading.seq_encoder.SeqFormat.qualityEncoding}\pysigline{\bfcode{qualityEncoding}}
If format comes with a known encoding of quality.


\strong{See also:}

\begin{description}
\item[{{\hyperref[fseq.reading:fseq.reading.seq_encoder.SeqEncoder.qualityEncoding]{\code{SeqEncoder.qualityEncoding}}}}] \leavevmode
Setter of encoder quality.

\end{description}



\end{fulllineitems}


\end{fulllineitems}

\index{SeqFormatDetector (class in fseq.reading.seq\_encoder)}

\begin{fulllineitems}
\phantomsection\label{fseq.reading:fseq.reading.seq_encoder.SeqFormatDetector}\pysiglinewithargsret{\strong{class }\code{fseq.reading.seq\_encoder.}\bfcode{SeqFormatDetector}}{\emph{forceFormat=None}}{}
Bases: \code{object}

Detection of data-format manager

Given a set of initially specified formats the detector feeds them lines
of data until only one remains \code{True}.
It then further continues a little while to be more certain that it
was not a mere fluke.


\strong{See also:}

\begin{description}
\item[{{\hyperref[fseq.reading:fseq.reading.seq_encoder.SeqFormat]{\code{SeqFormat}}}}] \leavevmode
Base class for formats that can be detected.

\end{description}


\paragraph{Attributes}

\begin{longtable}{ll}
\hline
\endfirsthead

\multicolumn{2}{c}%
{{\textsf{\tablename\ \thetable{} -- continued from previous page}}} \\
\hline
\endhead

\hline \multicolumn{2}{|r|}{{\textsf{Continued on next page}}} \\ \hline
\endfoot

\endlastfoot


{\hyperref[fseq.reading:fseq.reading.seq_encoder.SeqFormatDetector.detecting]{\code{detecting}}}
 & 
If attempting to detect: bool
\\
\hline
{\hyperref[fseq.reading:fseq.reading.seq_encoder.SeqFormatDetector.format]{\code{format}}}
 & 
The format name of the detected format: str
\\
\hline
{\hyperref[fseq.reading:fseq.reading.seq_encoder.SeqFormatDetector.itemSize]{\code{itemSize}}}
 & 
The size (lines) of each entry: int
\\
\hline
{\hyperref[fseq.reading:fseq.reading.seq_encoder.SeqFormatDetector.hasSequence]{\code{hasSequence}}}
 & 
If format has sequence information: bool
\\
\hline
{\hyperref[fseq.reading:fseq.reading.seq_encoder.SeqFormatDetector.hasQuality]{\code{hasQuality}}}
 & 
If format quality information: bool
\\
\hline
{\hyperref[fseq.reading:fseq.reading.seq_encoder.SeqFormatDetector.qualityEncoding]{\code{qualityEncoding}}}
 & 
If format comes with a known encoding of quality.
\\
\hline
{\hyperref[fseq.reading:fseq.reading.seq_encoder.SeqFormatDetector.headerLine]{\code{headerLine}}}
 & 
The line index in the item for the header
\\
\hline
{\hyperref[fseq.reading:fseq.reading.seq_encoder.SeqFormatDetector.sequenceLine]{\code{sequenceLine}}}
 & 
The line index in the item for the sequence
\\
\hline
{\hyperref[fseq.reading:fseq.reading.seq_encoder.SeqFormatDetector.qualityLine]{\code{qualityLine}}}
 & 
The line index in the item for the quality
\\
\hline
{\hyperref[fseq.reading:fseq.reading.seq_encoder.SeqFormatDetector.headerLine]{\code{headerLine}}}
 & 
The line index in the item for the header
\\
\hline
{\hyperref[fseq.reading:fseq.reading.seq_encoder.SeqFormatDetector.sequenceLine]{\code{sequenceLine}}}
 & 
The line index in the item for the sequence
\\
\hline
{\hyperref[fseq.reading:fseq.reading.seq_encoder.SeqFormatDetector.qualityLine]{\code{qualityLine}}}
 & 
The line index in the item for the quality
\\
\hline\end{longtable}


\begin{tabulary}{\linewidth}{|L|L|}
\hline

FORMATS
 & \\
\hline\end{tabulary}

\index{FORMATS (fseq.reading.seq\_encoder.SeqFormatDetector attribute)}

\begin{fulllineitems}
\phantomsection\label{fseq.reading:fseq.reading.seq_encoder.SeqFormatDetector.FORMATS}\pysigline{\bfcode{FORMATS}\strong{ = {[}\textless{}class `fseq.reading.seq\_encoder.FastaSingleline'\textgreater{}, \textless{}class `fseq.reading.seq\_encoder.FastaMultiline'\textgreater{}, \textless{}class `fseq.reading.seq\_encoder.FastQ'\textgreater{}{]}}}
\end{fulllineitems}

\index{compatible() (fseq.reading.seq\_encoder.SeqFormatDetector method)}

\begin{fulllineitems}
\phantomsection\label{fseq.reading:fseq.reading.seq_encoder.SeqFormatDetector.compatible}\pysiglinewithargsret{\bfcode{compatible}}{\emph{encoder}}{}
Evaluates if encoder is compatible with the detected format
\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
bool
\begin{quote}

Compatibility
\end{quote}

\item[{Raises}] \leavevmode
\textbf{FormatError}
\begin{quote}

If attempting to test compatibility before format is detected
\end{quote}

\end{description}\end{quote}

\end{fulllineitems}

\index{detecting (fseq.reading.seq\_encoder.SeqFormatDetector attribute)}

\begin{fulllineitems}
\phantomsection\label{fseq.reading:fseq.reading.seq_encoder.SeqFormatDetector.detecting}\pysigline{\bfcode{detecting}}
If attempting to detect: bool

\end{fulllineitems}

\index{feed() (fseq.reading.seq\_encoder.SeqFormatDetector method)}

\begin{fulllineitems}
\phantomsection\label{fseq.reading:fseq.reading.seq_encoder.SeqFormatDetector.feed}\pysiglinewithargsret{\bfcode{feed}}{\emph{line}}{}
Supply a new line to format detector.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\textbf{line: str}
\begin{quote}

The next line in the data
\end{quote}

\item[{Returns}] \leavevmode
fseq.SeqEncoder
\begin{quote}

Returns \code{self}
\end{quote}

\item[{Raises}] \leavevmode
\textbf{FormatUnknown}
\begin{quote}

If no known formatters are left
\end{quote}

\end{description}\end{quote}

\end{fulllineitems}

\index{format (fseq.reading.seq\_encoder.SeqFormatDetector attribute)}

\begin{fulllineitems}
\phantomsection\label{fseq.reading:fseq.reading.seq_encoder.SeqFormatDetector.format}\pysigline{\bfcode{format}}
The format name of the detected format: str

\end{fulllineitems}

\index{hasQuality (fseq.reading.seq\_encoder.SeqFormatDetector attribute)}

\begin{fulllineitems}
\phantomsection\label{fseq.reading:fseq.reading.seq_encoder.SeqFormatDetector.hasQuality}\pysigline{\bfcode{hasQuality}}
If format quality information: bool

\end{fulllineitems}

\index{hasSequence (fseq.reading.seq\_encoder.SeqFormatDetector attribute)}

\begin{fulllineitems}
\phantomsection\label{fseq.reading:fseq.reading.seq_encoder.SeqFormatDetector.hasSequence}\pysigline{\bfcode{hasSequence}}
If format has sequence information: bool

\end{fulllineitems}

\index{headerLine (fseq.reading.seq\_encoder.SeqFormatDetector attribute)}

\begin{fulllineitems}
\phantomsection\label{fseq.reading:fseq.reading.seq_encoder.SeqFormatDetector.headerLine}\pysigline{\bfcode{headerLine}}
The line index in the item for the header
\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
int

\item[{Raises}] \leavevmode
\textbf{FormatError}
\begin{quote}

If attempting to use before format is detected
\end{quote}

\end{description}\end{quote}

\end{fulllineitems}

\index{itemSize (fseq.reading.seq\_encoder.SeqFormatDetector attribute)}

\begin{fulllineitems}
\phantomsection\label{fseq.reading:fseq.reading.seq_encoder.SeqFormatDetector.itemSize}\pysigline{\bfcode{itemSize}}
The size (lines) of each entry: int

\end{fulllineitems}

\index{qualityEncoding (fseq.reading.seq\_encoder.SeqFormatDetector attribute)}

\begin{fulllineitems}
\phantomsection\label{fseq.reading:fseq.reading.seq_encoder.SeqFormatDetector.qualityEncoding}\pysigline{\bfcode{qualityEncoding}}
If format comes with a known encoding of quality.


\strong{See also:}

\begin{description}
\item[{{\hyperref[fseq.reading:fseq.reading.seq_encoder.SeqEncoder.qualityEncoding]{\code{SeqEncoder.qualityEncoding}}}}] \leavevmode
Setter of encoder quality.

\end{description}



\end{fulllineitems}

\index{qualityLine (fseq.reading.seq\_encoder.SeqFormatDetector attribute)}

\begin{fulllineitems}
\phantomsection\label{fseq.reading:fseq.reading.seq_encoder.SeqFormatDetector.qualityLine}\pysigline{\bfcode{qualityLine}}
The line index in the item for the quality
\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
int

\item[{Raises}] \leavevmode
\textbf{FormatError}
\begin{quote}

If attempting to use before format is detected
\end{quote}

\end{description}\end{quote}

\end{fulllineitems}

\index{sequenceLine (fseq.reading.seq\_encoder.SeqFormatDetector attribute)}

\begin{fulllineitems}
\phantomsection\label{fseq.reading:fseq.reading.seq_encoder.SeqFormatDetector.sequenceLine}\pysigline{\bfcode{sequenceLine}}
The line index in the item for the sequence
\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
int

\item[{Raises}] \leavevmode
\textbf{FormatError}
\begin{quote}

If attempting to use before format is detected
\end{quote}

\end{description}\end{quote}

\end{fulllineitems}


\end{fulllineitems}

\index{inheritDocFromSeqFormat() (in module fseq.reading.seq\_encoder)}

\begin{fulllineitems}
\phantomsection\label{fseq.reading:fseq.reading.seq_encoder.inheritDocFromSeqFormat}\pysiglinewithargsret{\code{fseq.reading.seq\_encoder.}\bfcode{inheritDocFromSeqFormat}}{\emph{f}}{}
This decorator will copy the docstring from \code{SeqFormat} for the
matching function \code{f} if such exists and no docstring has been added
manually.

\end{fulllineitems}



\subparagraph{fseq.reading.seq\_reader module}
\label{fseq.reading:module-fseq.reading.seq_reader}\label{fseq.reading:fseq-reading-seq-reader-module}\index{fseq.reading.seq\_reader (module)}
Module for reading sequence data
\index{SeqReader (class in fseq.reading.seq\_reader)}

\begin{fulllineitems}
\phantomsection\label{fseq.reading:fseq.reading.seq_reader.SeqReader}\pysiglinewithargsret{\strong{class }\code{fseq.reading.seq\_reader.}\bfcode{SeqReader}}{\emph{seqEncoder=None}, \emph{dataSourcePaths=None}, \emph{dataTargetPaths=None}, \emph{reportBuilders=None}, \emph{popDataSources=True}, \emph{resetSeqEncoder=True}, \emph{popEncodingResults=None}, \emph{dataArrayConstructor=\textless{}built-in function zeros\textgreater{}}, \emph{dataWidth=101}, \emph{dataType=\textless{}type `numpy.float16'\textgreater{}}, \emph{verbose=False}}{}
Bases: \code{object}

Reads sequence data and encodes it.

The length of the sequence reader reflects the number of inputs to be
processed.

The entire stack of inputs can be processed in bulk by invoking
\code{SeqReader.run()} but the results of each encoding, omitting any
report building can also be produced iteratively as shown in the examples.
\paragraph{Examples}

To invoke the reader it can either be run :

\begin{Verbatim}[commandchars=\\\{\}]
\PYG{g+gp}{\PYGZgt{}\PYGZgt{}\PYGZgt{} }\PYG{n}{seqReader}\PYG{o}{.}\PYG{n}{run}\PYG{p}{(}\PYG{p}{)}
\PYG{g+go}{\PYGZlt{}fseq.reading.seq\PYGZus{}reader.SeqReader at 0x7faf6970fd10\PYGZgt{}}
\end{Verbatim}

or if more control is required, it can be iterated over:

\begin{Verbatim}[commandchars=\\\{\}]
\PYG{g+gp}{\PYGZgt{}\PYGZgt{}\PYGZgt{} }\PYG{k}{for} \PYG{n}{res} \PYG{o+ow}{in} \PYG{n}{seqReader}\PYG{p}{:}
\PYG{g+gp}{... }   \PYG{n}{reportBuilder}\PYG{o}{.}\PYG{n}{distill}\PYG{p}{(}\PYG{n}{res}\PYG{p}{,} \PYG{n}{dirname}\PYG{o}{=}\PYG{n}{seqReader}\PYG{o}{.}\PYG{n}{reportDirectory}\PYG{p}{)}
\end{Verbatim}

Both methods above yielding the same result with the difference that
the first may store results in \code{seqReader.results} depending on the
\code{seqReader.popEncodingResults} settings while the latter never keeps
the results in the state of the instance.
\paragraph{Attributes}

\begin{longtable}{ll}
\hline
\endfirsthead

\multicolumn{2}{c}%
{{\textsf{\tablename\ \thetable{} -- continued from previous page}}} \\
\hline
\endhead

\hline \multicolumn{2}{|r|}{{\textsf{Continued on next page}}} \\ \hline
\endfoot

\endlastfoot


{\hyperref[fseq.reading:fseq.reading.seq_reader.SeqReader.popDataSources]{\code{popDataSources}}}
 & 
If sequence reader should remove data sources from list of sources when they have been read.
\\
\hline
{\hyperref[fseq.reading:fseq.reading.seq_reader.SeqReader.popEncodingResults]{\code{popEncodingResults}}}
 & 
If the outcome of an encoding should be remove from memory as soon as report as been produced or iteration completed.
\\
\hline
{\hyperref[fseq.reading:fseq.reading.seq_reader.SeqReader.jobQueue]{\code{jobQueue}}}
 & 
The data sources to be read and their respective targets.
\\
\hline
{\hyperref[fseq.reading:fseq.reading.seq_reader.SeqReader.reportBuilders]{\code{reportBuilders}}}
 & 
The report builders associated with the reader.
\\
\hline
{\hyperref[fseq.reading:fseq.reading.seq_reader.SeqReader.reportDirectory]{\code{reportDirectory}}}
 & 
The directory where reports for the last made encoding should go.
\\
\hline
{\hyperref[fseq.reading:fseq.reading.seq_reader.SeqReader.resetSeqEncoder]{\code{resetSeqEncoder}}}
 & 
If each data source will detect format anew or if all files are
\\
\hline
{\hyperref[fseq.reading:fseq.reading.seq_reader.SeqReader.results]{\code{results}}}
 & 
A list of the results of the encodings.
\\
\hline\end{longtable}


\begin{tabulary}{\linewidth}{|L|L|}
\hline

dataArrayConstructor
 & \\
\hline
dataWidth
 & \\
\hline
dataType
 & \\
\hline
SeqEncoder
 & \\
\hline\end{tabulary}

\index{DATA\_INITIAL\_SIZE (fseq.reading.seq\_reader.SeqReader attribute)}

\begin{fulllineitems}
\phantomsection\label{fseq.reading:fseq.reading.seq_reader.SeqReader.DATA_INITIAL_SIZE}\pysigline{\bfcode{DATA\_INITIAL\_SIZE}\strong{ = 100000}}
\end{fulllineitems}

\index{DEBUG (fseq.reading.seq\_reader.SeqReader attribute)}

\begin{fulllineitems}
\phantomsection\label{fseq.reading:fseq.reading.seq_reader.SeqReader.DEBUG}\pysigline{\bfcode{DEBUG}\strong{ = False}}
\end{fulllineitems}

\index{WORKERS (fseq.reading.seq\_reader.SeqReader attribute)}

\begin{fulllineitems}
\phantomsection\label{fseq.reading:fseq.reading.seq_reader.SeqReader.WORKERS}\pysigline{\bfcode{WORKERS}\strong{ = 32}}
\end{fulllineitems}

\index{addData() (fseq.reading.seq\_reader.SeqReader method)}

\begin{fulllineitems}
\phantomsection\label{fseq.reading:fseq.reading.seq_reader.SeqReader.addData}\pysiglinewithargsret{\bfcode{addData}}{\emph{sourcePaths}, \emph{targetPaths=None}}{}
Add a data source path to be analysed.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\textbf{sourcePaths: string or iterable object}
\begin{quote}

Either a path string or a collection of paths to data files
\end{quote}

\textbf{targetPaths: string or iterable object, optional}
\begin{quote}

Either a relative path string or collection of relative paths.
The path is relative to the respective data source.

(Default: will create a folder in the same directory as the 
data source with the same name as the input suffixed by
\code{.reports}.)
\begin{description}
\item[{\textbf{Note:} If supplied, must reflect equal number of outputs as}] \leavevmode
inputs in \code{sourcePaths}

\end{description}
\end{quote}

\item[{Returns}] \leavevmode
fseq.SeqReader
\begin{quote}

Returns \code{self}
\end{quote}

\item[{Raises}] \leavevmode
\textbf{ValueError}
\begin{quote}

If target paths are supplied but don't match in lenght with the
number of source-paths
\end{quote}

\end{description}\end{quote}

\end{fulllineitems}

\index{addReportBuilders() (fseq.reading.seq\_reader.SeqReader method)}

\begin{fulllineitems}
\phantomsection\label{fseq.reading:fseq.reading.seq_reader.SeqReader.addReportBuilders}\pysiglinewithargsret{\bfcode{addReportBuilders}}{\emph{*reportBuilders}}{}
Add a report builder to the set of reports done upon analysis.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\textbf{reportBuilders: fseq.ReportBuilder, optional}
\begin{quote}

Any number of report builder to be added
\end{quote}

\item[{Returns}] \leavevmode
fseq.SeqReader
\begin{quote}

Returns \code{self}
\end{quote}

\item[{Raises}] \leavevmode
\textbf{TypeError}
\begin{quote}

If an item in \code{reportBuilders} is not a valid
\code{fseq.ReportBuilder}
\end{quote}

\end{description}\end{quote}

\end{fulllineitems}

\index{clearJobQueue() (fseq.reading.seq\_reader.SeqReader method)}

\begin{fulllineitems}
\phantomsection\label{fseq.reading:fseq.reading.seq_reader.SeqReader.clearJobQueue}\pysiglinewithargsret{\bfcode{clearJobQueue}}{}{}
Removes all jobs in the queue.
\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
fseq.SeqReader
\begin{quote}

Returns \code{self}
\end{quote}

\end{description}\end{quote}

\end{fulllineitems}

\index{clearResults() (fseq.reading.seq\_reader.SeqReader method)}

\begin{fulllineitems}
\phantomsection\label{fseq.reading:fseq.reading.seq_reader.SeqReader.clearResults}\pysiglinewithargsret{\bfcode{clearResults}}{}{}
Removes all stored results of encodings
\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
fseq.SeqReader
\begin{quote}

Returns \code{self}
\end{quote}

\end{description}\end{quote}

\end{fulllineitems}

\index{dataArrayConstructor (fseq.reading.seq\_reader.SeqReader attribute)}

\begin{fulllineitems}
\phantomsection\label{fseq.reading:fseq.reading.seq_reader.SeqReader.dataArrayConstructor}\pysigline{\bfcode{dataArrayConstructor}}
\end{fulllineitems}

\index{dataType (fseq.reading.seq\_reader.SeqReader attribute)}

\begin{fulllineitems}
\phantomsection\label{fseq.reading:fseq.reading.seq_reader.SeqReader.dataType}\pysigline{\bfcode{dataType}}
\end{fulllineitems}

\index{dataWidth (fseq.reading.seq\_reader.SeqReader attribute)}

\begin{fulllineitems}
\phantomsection\label{fseq.reading:fseq.reading.seq_reader.SeqReader.dataWidth}\pysigline{\bfcode{dataWidth}}
\end{fulllineitems}

\index{jobQueue (fseq.reading.seq\_reader.SeqReader attribute)}

\begin{fulllineitems}
\phantomsection\label{fseq.reading:fseq.reading.seq_reader.SeqReader.jobQueue}\pysigline{\bfcode{jobQueue}}
The data sources to be read and their respective targets.
\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
list of tuples
\begin{quote}

Each tuple representing a source - target pair.
\end{quote}

\end{description}\end{quote}

\end{fulllineitems}

\index{next() (fseq.reading.seq\_reader.SeqReader method)}

\begin{fulllineitems}
\phantomsection\label{fseq.reading:fseq.reading.seq_reader.SeqReader.next}\pysiglinewithargsret{\bfcode{next}}{}{}
Part of iter interface, produces the encoding of the next
data-source.
\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
numpy.ndarray
\begin{quote}

Encoding output.
\end{quote}

\item[{Raises}] \leavevmode
\textbf{ValueError}
\begin{quote}

If no encoder has been assigned.
\end{quote}

\textbf{StopIteration}
\begin{quote}

If no more data-source exists.
\end{quote}

\end{description}\end{quote}

\end{fulllineitems}

\index{popDataSources (fseq.reading.seq\_reader.SeqReader attribute)}

\begin{fulllineitems}
\phantomsection\label{fseq.reading:fseq.reading.seq_reader.SeqReader.popDataSources}\pysigline{\bfcode{popDataSources}}
If sequence reader should remove data sources from list of sources
when they have been read.

The general use case is to set this to \code{True}, but omitting popping
can be useful if data may be needed to be re-read with a different
encoder.
\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
bool

\end{description}\end{quote}

\end{fulllineitems}

\index{popEncodingResults (fseq.reading.seq\_reader.SeqReader attribute)}

\begin{fulllineitems}
\phantomsection\label{fseq.reading:fseq.reading.seq_reader.SeqReader.popEncodingResults}\pysigline{\bfcode{popEncodingResults}}
If the outcome of an encoding should be remove from memory as soon
as report as been produced or iteration completed.

\textbf{Note:} The process will require large amounts of memory if results
are not popped and several large files analysed.
\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
bool

\end{description}\end{quote}

\end{fulllineitems}

\index{removeReportBuilders() (fseq.reading.seq\_reader.SeqReader method)}

\begin{fulllineitems}
\phantomsection\label{fseq.reading:fseq.reading.seq_reader.SeqReader.removeReportBuilders}\pysiglinewithargsret{\bfcode{removeReportBuilders}}{\emph{*builders}}{}
Removes all builders supplied, or all builders if no specific
builder is supplied.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\textbf{*args: fseq.ReportBuilder, optional}
\begin{quote}

Any number of report builder references for report builders to
be removed.
If none is supplied, all report builders will be removed.
\end{quote}

\item[{Returns}] \leavevmode
fseq.SeqReader
\begin{quote}

Returns \code{self}
\end{quote}

\end{description}\end{quote}
\paragraph{Examples}

If current instance has three builders:

\begin{Verbatim}[commandchars=\\\{\}]
\PYG{g+gp}{\PYGZgt{}\PYGZgt{}\PYGZgt{} }\PYG{n+nb}{tuple}\PYG{p}{(}\PYG{n}{seqEncoder}\PYG{o}{.}\PYG{n}{reportBuilders}\PYG{p}{)}
\PYG{g+go}{(b1, b2, b3)}
\end{Verbatim}

The \code{b2} and \code{b3} can be removed by:

\begin{Verbatim}[commandchars=\\\{\}]
\PYG{g+gp}{\PYGZgt{}\PYGZgt{}\PYGZgt{} }\PYG{n}{seqEncoder}\PYG{o}{.}\PYG{n}{removeReportBuilders}\PYG{p}{(}\PYG{n}{b2}\PYG{p}{,} \PYG{n}{b3}\PYG{p}{)}
\PYG{g+go}{\PYGZlt{}fseq.reading.seq\PYGZus{}reader.SeqReader at 0x7faf696e5810\PYGZgt{}}
\end{Verbatim}

\begin{Verbatim}[commandchars=\\\{\}]
\PYG{g+gp}{\PYGZgt{}\PYGZgt{}\PYGZgt{} }\PYG{n+nb}{tuple}\PYG{p}{(}\PYG{n}{seqEncoder}\PYG{o}{.}\PYG{n}{reportBuilders}\PYG{p}{)}
\PYG{g+go}{(b1, )}
\end{Verbatim}

Alternatively all builders can be removed by:

\begin{Verbatim}[commandchars=\\\{\}]
\PYG{g+gp}{\PYGZgt{}\PYGZgt{}\PYGZgt{} }\PYG{n}{seqEncoder}\PYG{o}{.}\PYG{n}{removeReportBuilders}\PYG{p}{(}\PYG{p}{)}
\PYG{g+go}{\PYGZlt{}fseq.reading.seq\PYGZus{}reader.SeqReader at 0x7faf696e5810\PYGZgt{}}
\end{Verbatim}

\begin{Verbatim}[commandchars=\\\{\}]
\PYG{g+gp}{\PYGZgt{}\PYGZgt{}\PYGZgt{} }\PYG{n+nb}{tuple}\PYG{p}{(}\PYG{n}{seqEncoder}\PYG{o}{.}\PYG{n}{reportBuilders}\PYG{p}{)}
\PYG{g+go}{()}
\end{Verbatim}

\end{fulllineitems}

\index{reportBuilders (fseq.reading.seq\_reader.SeqReader attribute)}

\begin{fulllineitems}
\phantomsection\label{fseq.reading:fseq.reading.seq_reader.SeqReader.reportBuilders}\pysigline{\bfcode{reportBuilders}}
The report builders associated with the reader.

If any, the reports will be destilled automatically at the end of
\code{SeqReader.run()}.
\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
tuple
\begin{quote}

The currently assigned report builders
\end{quote}

\end{description}\end{quote}


\strong{See also:}

\begin{description}
\item[{\code{fseq.reporting.report\_builder.ReportBuilder.distill}}] \leavevmode
Method for distilling encoded data.

\end{description}



\end{fulllineitems}

\index{reportDirectory (fseq.reading.seq\_reader.SeqReader attribute)}

\begin{fulllineitems}
\phantomsection\label{fseq.reading:fseq.reading.seq_reader.SeqReader.reportDirectory}\pysigline{\bfcode{reportDirectory}}
The directory where reports for the last made encoding should go.
\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
str

\end{description}\end{quote}

\end{fulllineitems}

\index{resetSeqEncoder (fseq.reading.seq\_reader.SeqReader attribute)}

\begin{fulllineitems}
\phantomsection\label{fseq.reading:fseq.reading.seq_reader.SeqReader.resetSeqEncoder}\pysigline{\bfcode{resetSeqEncoder}}
If each data source will detect format anew or if all files are
assumed to be of the same format
\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
bool

\end{description}\end{quote}

\end{fulllineitems}

\index{results (fseq.reading.seq\_reader.SeqReader attribute)}

\begin{fulllineitems}
\phantomsection\label{fseq.reading:fseq.reading.seq_reader.SeqReader.results}\pysigline{\bfcode{results}}
A list of the results of the encodings.
\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
list

\end{description}\end{quote}


\strong{See also:}

\begin{description}
\item[{\code{SeqEncoder.clearResults}}] \leavevmode
Clearing the list of results

\end{description}



\end{fulllineitems}

\index{run() (fseq.reading.seq\_reader.SeqReader method)}

\begin{fulllineitems}
\phantomsection\label{fseq.reading:fseq.reading.seq_reader.SeqReader.run}\pysiglinewithargsret{\bfcode{run}}{}{}
Runs through all sources and produces reports if such have been
attached.
\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
fseq.SeqReader
\begin{quote}

Returns \code{self}
\end{quote}

\end{description}\end{quote}

\end{fulllineitems}

\index{seqEncoder (fseq.reading.seq\_reader.SeqReader attribute)}

\begin{fulllineitems}
\phantomsection\label{fseq.reading:fseq.reading.seq_reader.SeqReader.seqEncoder}\pysigline{\bfcode{seqEncoder}}
The encoder attached to the sequence reader that will parse the
input strings into values.
\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
fseq.SeqEncoder
\begin{quote}

Current encoder
\end{quote}

\item[{Raises}] \leavevmode
\textbf{TypeError}
\begin{quote}

If trying to assign object that is not a \code{fseq.SeqEncoder}
\end{quote}

\end{description}\end{quote}

\end{fulllineitems}

\index{verbose (fseq.reading.seq\_reader.SeqReader attribute)}

\begin{fulllineitems}
\phantomsection\label{fseq.reading:fseq.reading.seq_reader.SeqReader.verbose}\pysigline{\bfcode{verbose}}
\end{fulllineitems}


\end{fulllineitems}



\subparagraph{Module contents}
\label{fseq.reading:module-fseq.reading}\label{fseq.reading:module-contents}\index{fseq.reading (module)}
Reading-related modules of fseq.

The reading package contains of two modules: \emph{seq\_encoder} and \emph{seq\_reader}.

The reader contains the generic reader that coordinates actions and works as
the mainframe of \emph{fseq}.
This module should need no extensions to increase the funcitonality of the
\emph{fseq}.

The encoder-module contains both the different types of encoders available as
well as the format-detectors for variaous types of input formats.
Here further formats can be added and new encoders written to extend the 
functionality of \emph{fseq}.


\paragraph{fseq.reporting package}
\label{fseq.reporting::doc}\label{fseq.reporting:fseq-reporting-package}

\subparagraph{Submodules}
\label{fseq.reporting:submodules}

\subparagraph{fseq.reporting.report\_builder module}
\label{fseq.reporting:module-fseq.reporting.report_builder}\label{fseq.reporting:fseq-reporting-report-builder-module}\index{fseq.reporting.report\_builder (module)}
The report builders are classes that coordinate report productions
\index{ReportBuilderBase (class in fseq.reporting.report\_builder)}

\begin{fulllineitems}
\phantomsection\label{fseq.reporting:fseq.reporting.report_builder.ReportBuilderBase}\pysiglinewithargsret{\strong{class }\code{fseq.reporting.report\_builder.}\bfcode{ReportBuilderBase}}{\emph{*reports}, \emph{**kwargs}}{}
Bases: \code{object}

Base class for common report builder features.

Most prominently, the distill-method needs to be implemented in
subclasses to make much sence in most use cases.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\textbf{outputRoot: str, optional}
\begin{quote}

Path to the directory where all reports should be put

(Default: \code{None})
\end{quote}

\textbf{outputNamePrefix: str, optional}
\begin{quote}

Partial name to be added to all reports done by the builder

(Default: \code{None})
\end{quote}

\textbf{*reports: objects, optional}
\begin{quote}

Any number of reports to be added from start
\end{quote}

\end{description}\end{quote}
\paragraph{Attributes}

\begin{longtable}{ll}
\hline
\endfirsthead

\multicolumn{2}{c}%
{{\textsf{\tablename\ \thetable{} -- continued from previous page}}} \\
\hline
\endhead

\hline \multicolumn{2}{|r|}{{\textsf{Continued on next page}}} \\ \hline
\endfoot

\endlastfoot


{\hyperref[fseq.reporting:fseq.reporting.report_builder.ReportBuilderBase.outputRoot]{\code{outputRoot}}}
 & 
The base for saving out reports: str
\\
\hline
{\hyperref[fseq.reporting:fseq.reporting.report_builder.ReportBuilderBase.outputNamePrefix]{\code{outputNamePrefix}}}
 & 
Partial file name to prepend the individual reports: str
\\
\hline\end{longtable}


\begin{tabulary}{\linewidth}{|L|L|}
\hline

DEFAULT\_REPORTS
 & \\
\hline\end{tabulary}

\index{DEFAULT\_REPORTS (fseq.reporting.report\_builder.ReportBuilderBase attribute)}

\begin{fulllineitems}
\phantomsection\label{fseq.reporting:fseq.reporting.report_builder.ReportBuilderBase.DEFAULT_REPORTS}\pysigline{\bfcode{DEFAULT\_REPORTS}\strong{ = ()}}
\end{fulllineitems}

\index{addReports() (fseq.reporting.report\_builder.ReportBuilderBase method)}

\begin{fulllineitems}
\phantomsection\label{fseq.reporting:fseq.reporting.report_builder.ReportBuilderBase.addReports}\pysiglinewithargsret{\bfcode{addReports}}{\emph{*reports}}{}
Adds any number of reports given that the reports exposes a 
distill method.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\textbf{*reports: objects, optional}

\item[{Returns}] \leavevmode
fseq.ReportBuilderBase
\begin{quote}

Returns \code{self}
\end{quote}

\item[{Raises}] \leavevmode
\textbf{ValueError}
\begin{quote}

If a report lacks a method named \code{distill} or can't be hashed
\end{quote}

\end{description}\end{quote}

\end{fulllineitems}

\index{distill() (fseq.reporting.report\_builder.ReportBuilderBase method)}

\begin{fulllineitems}
\phantomsection\label{fseq.reporting:fseq.reporting.report_builder.ReportBuilderBase.distill}\pysiglinewithargsret{\bfcode{distill}}{\emph{*args}, \emph{**kwargs}}{}
The base distiller will not process any data passed to it,
it will send all arguments and keyword arguments to the individual
reports.

If either \code{outputRoot} or \code{outputNamePrefix} are passed as kwargs,
the corresponding values preset in the system will be added to the
kwargs sent to the subreports.
\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
fseq.ReportBuilderBase
\begin{quote}

Returns \code{self}
\end{quote}

\end{description}\end{quote}

\end{fulllineitems}

\index{outputNamePrefix (fseq.reporting.report\_builder.ReportBuilderBase attribute)}

\begin{fulllineitems}
\phantomsection\label{fseq.reporting:fseq.reporting.report_builder.ReportBuilderBase.outputNamePrefix}\pysigline{\bfcode{outputNamePrefix}}
Partial file name to prepend the individual reports: str

\end{fulllineitems}

\index{outputRoot (fseq.reporting.report\_builder.ReportBuilderBase attribute)}

\begin{fulllineitems}
\phantomsection\label{fseq.reporting:fseq.reporting.report_builder.ReportBuilderBase.outputRoot}\pysigline{\bfcode{outputRoot}}
The base for saving out reports: str

\end{fulllineitems}


\end{fulllineitems}

\index{ReportBuilderFFT (class in fseq.reporting.report\_builder)}

\begin{fulllineitems}
\phantomsection\label{fseq.reporting:fseq.reporting.report_builder.ReportBuilderFFT}\pysiglinewithargsret{\strong{class }\code{fseq.reporting.report\_builder.}\bfcode{ReportBuilderFFT}}{\emph{*reports}, \emph{**kwargs}}{}
Bases: {\hyperref[fseq.reporting:fseq.reporting.report_builder.ReportBuilderBase]{\code{fseq.reporting.report\_builder.ReportBuilderBase}}}

Samples part of data set and performs FFT-based analysis on it.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\textbf{outputRoot: str, optional}
\begin{quote}

Path to the directory where all reports should be put

(Default: \code{None})
\end{quote}

\textbf{outputNamePrefix: str, optional}
\begin{quote}

Partial name to be added to all reports done by the builder

(Default: \code{None})
\end{quote}

\textbf{sampleSize: int, optional}
\begin{quote}

Size of sample to be randomly drawn

(Default: 1000)
\end{quote}

\textbf{distanceMetric: str, optional}
\begin{quote}

Name of distance metric to be used.
See \code{METRICS}-attribute for allowed metrics.

(Default: `correlation')
\end{quote}

\textbf{*reports: objects, optional}
\begin{quote}

Any number of reports to be added from start

(Default: A fseq.HeatMap)
\end{quote}

\end{description}\end{quote}


\strong{See also:}

\begin{description}
\item[{{\hyperref[fseq.reporting:fseq.reporting.report_builder.ReportBuilderBase]{\code{ReportBuilderBase}}}}] \leavevmode
Base class implementing more attributes.

\end{description}


\paragraph{Attributes}

\begin{longtable}{ll}
\hline
\endfirsthead

\multicolumn{2}{c}%
{{\textsf{\tablename\ \thetable{} -- continued from previous page}}} \\
\hline
\endhead

\hline \multicolumn{2}{|r|}{{\textsf{Continued on next page}}} \\ \hline
\endfoot

\endlastfoot


{\hyperref[fseq.reporting:fseq.reporting.report_builder.ReportBuilderFFT.distanceMetric]{\code{distanceMetric}}}
 & 
The ReportBuilderFFT.METRIC used: str
\\
\hline
{\hyperref[fseq.reporting:fseq.reporting.report_builder.ReportBuilderFFT.sampleSize]{\code{sampleSize}}}
 & 
Size of data subsample to analyze: int
\\
\hline\end{longtable}

\index{DEFAULT\_REPORTS (fseq.reporting.report\_builder.ReportBuilderFFT attribute)}

\begin{fulllineitems}
\phantomsection\label{fseq.reporting:fseq.reporting.report_builder.ReportBuilderFFT.DEFAULT_REPORTS}\pysigline{\bfcode{DEFAULT\_REPORTS}\strong{ = (\textless{}class `fseq.reporting.reports.HeatMap'\textgreater{},)}}
\end{fulllineitems}

\index{METRICS (fseq.reporting.report\_builder.ReportBuilderFFT attribute)}

\begin{fulllineitems}
\phantomsection\label{fseq.reporting:fseq.reporting.report_builder.ReportBuilderFFT.METRICS}\pysigline{\bfcode{METRICS}\strong{ = set({[}'kulsinski', `chebyshev', `yule', `sokalmichener', `dice', `canberra', `jaccard', `minkowski', `seuclidean', `mahalanobis', `sqeuclidean', `euclidean', `russellrao', `cosine', `sokalsneath', `braycurtis', `hamming', `correlation', `cityblock', `rogerstanimoto', `matching'{]})}}
\end{fulllineitems}

\index{distanceMetric (fseq.reporting.report\_builder.ReportBuilderFFT attribute)}

\begin{fulllineitems}
\phantomsection\label{fseq.reporting:fseq.reporting.report_builder.ReportBuilderFFT.distanceMetric}\pysigline{\bfcode{distanceMetric}}
The ReportBuilderFFT.METRIC used: str

\end{fulllineitems}

\index{distill() (fseq.reporting.report\_builder.ReportBuilderFFT method)}

\begin{fulllineitems}
\phantomsection\label{fseq.reporting:fseq.reporting.report_builder.ReportBuilderFFT.distill}\pysiglinewithargsret{\bfcode{distill}}{\emph{data}, \emph{distanceMetric=None}, \emph{clusterOnAbsOnly=True}, \emph{*args}, \emph{**kwargs}}{}
Make reports from data.

Produces two analyses:
\begin{description}
\item[{Amplitude evaluation}] \leavevmode
A clustered FFT-amplitude analysis

\item[{Angle evaluation}] \leavevmode
A clustered FFT-angle analysis

\end{description}
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\textbf{data: numpy.ndarray}
\begin{quote}

The 2D-array of data given
\end{quote}

\textbf{distanceMetric: str, optional}
\begin{quote}

A distance metric to overwrite the default one of the instance.

(Default: Value of \code{self.distanceMetric})
\end{quote}

\textbf{clusterOnAbsOnly: bool, optional}
\begin{quote}

If clustering should be performed only on the amplitude (abs-values)
or if amplitude and angle be clustered independently.

(Default: Cluster only on amplitude)
\end{quote}

\end{description}\end{quote}

\end{fulllineitems}

\index{sampleSize (fseq.reporting.report\_builder.ReportBuilderFFT attribute)}

\begin{fulllineitems}
\phantomsection\label{fseq.reporting:fseq.reporting.report_builder.ReportBuilderFFT.sampleSize}\pysigline{\bfcode{sampleSize}}
Size of data subsample to analyze: int

\end{fulllineitems}


\end{fulllineitems}

\index{ReportBuilderPositionAverage (class in fseq.reporting.report\_builder)}

\begin{fulllineitems}
\phantomsection\label{fseq.reporting:fseq.reporting.report_builder.ReportBuilderPositionAverage}\pysiglinewithargsret{\strong{class }\code{fseq.reporting.report\_builder.}\bfcode{ReportBuilderPositionAverage}}{\emph{*reports}, \emph{**kwargs}}{}
Bases: {\hyperref[fseq.reporting:fseq.reporting.report_builder.ReportBuilderBase]{\code{fseq.reporting.report\_builder.ReportBuilderBase}}}

Per position analysis builder.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\textbf{outputRoot: str, optional}
\begin{quote}

Path to the directory where all reports should be put

(Default: \code{None})
\end{quote}

\textbf{outputNamePrefix: str, optional}
\begin{quote}

Partial name to be added to all reports done by the builder

(Default: \code{None})
\end{quote}

\textbf{undecidedValue: int, optional}
\begin{quote}

The value for which undecided items were encoded (so it
can be omited and calculated separately for 2 of 3 graphs).

(Default: 0.5)
\end{quote}

\textbf{*reports: objects, optional}
\begin{quote}

Any number of reports to be added from start

(Default: fseq.LinePlot)
\end{quote}

\end{description}\end{quote}


\strong{See also:}

\begin{description}
\item[{{\hyperref[fseq.reporting:fseq.reporting.report_builder.ReportBuilderBase]{\code{ReportBuilderBase}}}}] \leavevmode
Base class which implements some more attributes.

\end{description}


\paragraph{Attributes}

\begin{tabulary}{\linewidth}{|L|L|}
\hline

undecidedValue
 & \\
\hline\end{tabulary}

\index{DEFAULT\_REPORTS (fseq.reporting.report\_builder.ReportBuilderPositionAverage attribute)}

\begin{fulllineitems}
\phantomsection\label{fseq.reporting:fseq.reporting.report_builder.ReportBuilderPositionAverage.DEFAULT_REPORTS}\pysigline{\bfcode{DEFAULT\_REPORTS}\strong{ = (\textless{}class `fseq.reporting.reports.LinePlot'\textgreater{},)}}
\end{fulllineitems}

\index{distill() (fseq.reporting.report\_builder.ReportBuilderPositionAverage method)}

\begin{fulllineitems}
\phantomsection\label{fseq.reporting:fseq.reporting.report_builder.ReportBuilderPositionAverage.distill}\pysiglinewithargsret{\bfcode{distill}}{\emph{data}, \emph{undecidedValue=None}, \emph{*args}, \emph{**kwargs}}{}
The distiller will create reports for several position-type
informations.
\begin{description}
\item[{Average Lacking Data Frequency}] \leavevmode
The number of undecided values.

\item[{Average Non-Lacking Data}] \leavevmode
The per position average for all non-lacking data values.

\item[{Average Combined Data}] \leavevmode
The per position average as encoded.

\end{description}
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\textbf{data: numpy.ndarray}
\begin{quote}

An array of numerically encoded sequence information
\end{quote}

\textbf{undecidedValue: float, optional}
\begin{quote}

The value that undecided sequence positions are encoded as.
If not supplied, the previously set value of the class instance
will be used.
(Default: 0.5)
\end{quote}

\textbf{*args:}
\begin{quote}

Any args will be passed to the \code{ReportBuilderBase.distill}
\end{quote}

\textbf{**kwargs:}
\begin{quote}

Any kwargs will be passed to the \code{ReportBuilderBase.distill}

\textbf{Note:} \code{outputNamePrefix} will be overwritten/added
\end{quote}

\item[{Returns}] \leavevmode
fseq.ReportBuilderPositionAverage
\begin{quote}

Returns \code{self}
\end{quote}

\end{description}\end{quote}

\end{fulllineitems}

\index{undecidedValue (fseq.reporting.report\_builder.ReportBuilderPositionAverage attribute)}

\begin{fulllineitems}
\phantomsection\label{fseq.reporting:fseq.reporting.report_builder.ReportBuilderPositionAverage.undecidedValue}\pysigline{\bfcode{undecidedValue}}
\end{fulllineitems}


\end{fulllineitems}



\subparagraph{fseq.reporting.reports module}
\label{fseq.reporting:module-fseq.reporting.reports}\label{fseq.reporting:fseq-reporting-reports-module}\index{fseq.reporting.reports (module)}
Module for holding the various implemented reporting classes
\index{HeatMap (class in fseq.reporting.reports)}

\begin{fulllineitems}
\phantomsection\label{fseq.reporting:fseq.reporting.reports.HeatMap}\pysiglinewithargsret{\strong{class }\code{fseq.reporting.reports.}\bfcode{HeatMap}}{\emph{name='heatmap.pdf'}, \emph{saveArgs=()}, \emph{saveKwargs=\{\}}}{}
Bases: {\hyperref[fseq.reporting:fseq.reporting.reports.ReportBase]{\code{fseq.reporting.reports.ReportBase}}}

Makes heatmaps from data
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\textbf{name: str, optional}
\begin{quote}

A specific name of the report

(Default: ``heatmap.pdf'')
\end{quote}

\textbf{saveArgs: tuple or list, optional}
\begin{quote}

Any args to be passed to \code{matplotlib.savefig} after the figure

(Default: Empty tuple)
\end{quote}

\textbf{saveKwargs: dict, optional}
\begin{quote}

Any keyword args to be passed to \code{matplotlib.savefig}

(Default: Empty dict)
\end{quote}

\end{description}\end{quote}
\index{distill() (fseq.reporting.reports.HeatMap method)}

\begin{fulllineitems}
\phantomsection\label{fseq.reporting:fseq.reporting.reports.HeatMap.distill}\pysiglinewithargsret{\bfcode{distill}}{\emph{data}, \emph{name=None}, \emph{outputRoot=None}, \emph{outputNamePrefix=None}, \emph{title=None}, \emph{text=None}, \emph{ylabel=None}, \emph{xlabel=None}, \emph{saveArgs=()}, \emph{saveKwargs=\{\}}, \emph{vmin=None}, \emph{vmax=None}, \emph{aspect='auto'}, \emph{axisOff=True}, \emph{cmap=\textless{}matplotlib.colors.LinearSegmentedColormap object at 0x2b9813ac4b10\textgreater{}}, \emph{*args}, \emph{**kwargs}}{}
Creates the actual heatmap.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\textbf{data: numpy.ndarray}
\begin{quote}

The data to be plotted
\end{quote}

\textbf{name: str, optional}
\begin{quote}

If the default name of the HeatMap instance should be overwritten

(Default: Use the value of \code{self.name})
\end{quote}

\textbf{outputRoot: str, optional}
\begin{quote}

The directory in which to place the report

(Default: \code{None})
\end{quote}

\textbf{outputNamePrefix: str, optional}
\begin{quote}

A prefix to prepend the name when saving the output.
Typically set by the builder to indicate what post-processing
was done to the data shown in the report.

(Default: \code{None})
\end{quote}

\textbf{title: str, optional}
\begin{quote}

A title to be put over the heatmap
(Default: \code{None}, value automatically scaled by matplotlib)

(Default: \code{None})
\end{quote}

\textbf{text: str, optional}
\begin{quote}

An explanatory text to put under the plot

(Default: \code{None})

\textbf{Note:} This feature has not been implemented yet.
\end{quote}

\textbf{ylabel: str, optional}
\begin{quote}

A label for the y-axis

(Default: \code{None})
\end{quote}

\textbf{xlabel: str, optional}
\begin{quote}

A label for the x-axis

(Default: \code{None})
\end{quote}

\textbf{saveArgs: tuple or list, optional}
\begin{quote}

A set of arguments to overwrite the instance's default save args

(Default: empty tuple)
\end{quote}

\textbf{saveKwargs: dict, optional}
\begin{quote}

A set of keyword arguments to overwrite the instanc's default

(Default: empty dict)
\end{quote}

\textbf{vmin: number, optional}
\begin{quote}

To set a minimum color-scale number for the heatmap

(Default: \code{None}, value automatically scaled by matplotlib)
\end{quote}

\textbf{vmax: number, optional}
\begin{quote}

To set a maximum color-scale number for the heatmap

(Default: \code{None}, value automatically scaled by matplotlib)
\end{quote}

\textbf{aspect: str, optional}
\begin{quote}

The aspect ratio of the blocks/pixels in the heatmap

(Default: `auto', which allows for rectangular pixels)
\end{quote}

\textbf{axisOff: bool, optional}
\begin{quote}

If the axis of the plot should not be rendered

(Default: True)
\end{quote}

\textbf{cmap: matplotlib.cmap, optional}
\begin{quote}

A colormap to be used when plotting.

(Default: Red -- Blue)
\end{quote}

\end{description}\end{quote}

\end{fulllineitems}


\end{fulllineitems}

\index{LinePlot (class in fseq.reporting.reports)}

\begin{fulllineitems}
\phantomsection\label{fseq.reporting:fseq.reporting.reports.LinePlot}\pysiglinewithargsret{\strong{class }\code{fseq.reporting.reports.}\bfcode{LinePlot}}{\emph{name='line.pdf'}, \emph{saveArgs=()}, \emph{saveKwargs=\{\}}}{}
Bases: {\hyperref[fseq.reporting:fseq.reporting.reports.ReportBase]{\code{fseq.reporting.reports.ReportBase}}}

Makes lines from data
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\textbf{name: str, optional}
\begin{quote}

A specific name of the report

(Default: ``line.pdf'')
\end{quote}

\textbf{saveArgs: tuple or list, optional}
\begin{quote}

Any args to be passed to \code{matplotlib.savefig} after the figure

(Default: Empty tuple)
\end{quote}

\textbf{saveKwargs: dict, optional}
\begin{quote}

Any keyword args to be passed to \code{matplotlib.savefig}

(Default: Empty dict)
\end{quote}

\end{description}\end{quote}
\index{distill() (fseq.reporting.reports.LinePlot method)}

\begin{fulllineitems}
\phantomsection\label{fseq.reporting:fseq.reporting.reports.LinePlot.distill}\pysiglinewithargsret{\bfcode{distill}}{\emph{data}, \emph{name=None}, \emph{outputRoot=None}, \emph{outputNamePrefix=None}, \emph{title=None}, \emph{text=None}, \emph{ylabel=None}, \emph{xlabel=None}, \emph{saveArgs=()}, \emph{saveKwargs=\{\}}, \emph{logX=False}, \emph{logY=False}, \emph{basex=None}, \emph{basey=None}, \emph{labels=None}, \emph{*args}, \emph{**kwargs}}{}
Creates the actual heatmap.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\textbf{data: numpy.ndarray}
\begin{quote}

The data to be plotted
\end{quote}

\textbf{name: str, optional}
\begin{quote}

If the default name of the HeatMap instance should be overwritten

(Default: Use the value of \code{self.name})
\end{quote}

\textbf{outputRoot: str, optional}
\begin{quote}

The directory in which to place the report

(Default: \code{None})
\end{quote}

\textbf{outputNamePrefix: str, optional}
\begin{quote}

A prefix to prepend the name when saving the output.
Typically set by the builder to indicate what post-processing
was done to the data shown in the report.

(Default: \code{None})
\end{quote}

\textbf{title: str, optional}
\begin{quote}

A title to be put over the heatmap
(Default: \code{None}, value automatically scaled by matplotlib)

(Default: \code{None})
\end{quote}

\textbf{text: str, optional}
\begin{quote}

An explanatory text to put under the plot

(Default: \code{None})

\textbf{Note:} This feature has not been implemented yet.
\end{quote}

\textbf{ylabel: str, optional}
\begin{quote}

A label for the y-axis

(Default: \code{None})
\end{quote}

\textbf{xlabel: str, optional}
\begin{quote}

A label for the x-axis

(Default: \code{None})
\end{quote}

\textbf{saveArgs: tuple or list, optional}
\begin{quote}

A set of arguments to overwrite the instance's default save args

(Default: empty \code{tuple})
\end{quote}

\textbf{saveKwargs: dict, optional}
\begin{quote}

A set of keyword arguments to overwrite the instanc's default

(Default: empty \code{dict})
\end{quote}

\textbf{logX: bool, optional}
\begin{quote}

If X-axis should be logged

(Default: \code{False})
\end{quote}

\textbf{logY: bool, optional}
\begin{quote}

If Y-axis should be logged

(Default: \code{False})
\end{quote}

\textbf{basex: number, optional}
\begin{quote}

To specify other then 10-base logging

(Default: \code{None}, uses 10-base)
\end{quote}

\textbf{basey: number, optional}
\begin{quote}

To specify other than 10-base logging

(Default: \code{None}, uses 10-base)
\end{quote}

\textbf{labels: str, optional}
\begin{quote}

To name the line plotted and thus add a legend to the plot.

(Default: \code{None})
\end{quote}

\end{description}\end{quote}

\end{fulllineitems}


\end{fulllineitems}

\index{ReportBase (class in fseq.reporting.reports)}

\begin{fulllineitems}
\phantomsection\label{fseq.reporting:fseq.reporting.reports.ReportBase}\pysiglinewithargsret{\strong{class }\code{fseq.reporting.reports.}\bfcode{ReportBase}}{\emph{name=None}, \emph{saveArgs=()}, \emph{saveKwargs=\{\}}}{}
Bases: \code{object}

Base class for simple report creations.

Main purpose is to make a common interface for figure saving using
matplotlib figures.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\textbf{name: str, optional}
\begin{quote}

A specific name of the report

(Default: \code{None})
\end{quote}

\textbf{saveArgs: tuple or list, optional}
\begin{quote}

Any args to be passed to \code{matplotlib.savefig} after the figure

(Default: Empty \code{tuple})
\end{quote}

\textbf{saveKwargs: dict, optional}
\begin{quote}

Any keyword args to be passed to \code{matplotlib.savefig}

(Default: Empty \code{dict})
\end{quote}

\end{description}\end{quote}
\paragraph{Attributes}

\begin{longtable}{ll}
\hline
\endfirsthead

\multicolumn{2}{c}%
{{\textsf{\tablename\ \thetable{} -- continued from previous page}}} \\
\hline
\endhead

\hline \multicolumn{2}{|r|}{{\textsf{Continued on next page}}} \\ \hline
\endfoot

\endlastfoot


{\hyperref[fseq.reporting:fseq.reporting.reports.ReportBase.name]{\code{name}}}
 & 
Name of the plot, used to name the file: str
\\
\hline
{\hyperref[fseq.reporting:fseq.reporting.reports.ReportBase.saveArgs]{\code{saveArgs}}}
 & 
Save args passed to \code{matplotlib.pyplot.figure.savefig}: tuple
\\
\hline
{\hyperref[fseq.reporting:fseq.reporting.reports.ReportBase.saveKwargs]{\code{saveKwargs}}}
 & 
Save keyword args passed to \code{matplotlib.pyplot.figure.savefig}:
\\
\hline\end{longtable}

\index{distill() (fseq.reporting.reports.ReportBase method)}

\begin{fulllineitems}
\phantomsection\label{fseq.reporting:fseq.reporting.reports.ReportBase.distill}\pysiglinewithargsret{\bfcode{distill}}{\emph{data}, \emph{outputRoot=None}, \emph{outputNamePrefix=None}, \emph{*args}, \emph{**kwargs}}{}
Placeholder distill interface not to be used.
\begin{quote}\begin{description}
\item[{Raises}] \leavevmode
\textbf{NotImplementedError}
\begin{quote}

Always raises this exception
\end{quote}

\end{description}\end{quote}

\end{fulllineitems}

\index{name (fseq.reporting.reports.ReportBase attribute)}

\begin{fulllineitems}
\phantomsection\label{fseq.reporting:fseq.reporting.reports.ReportBase.name}\pysigline{\bfcode{name}}
Name of the plot, used to name the file: str

\end{fulllineitems}

\index{saveArgs (fseq.reporting.reports.ReportBase attribute)}

\begin{fulllineitems}
\phantomsection\label{fseq.reporting:fseq.reporting.reports.ReportBase.saveArgs}\pysigline{\bfcode{saveArgs}}
Save args passed to \code{matplotlib.pyplot.figure.savefig}: tuple

\end{fulllineitems}

\index{saveFig() (fseq.reporting.reports.ReportBase method)}

\begin{fulllineitems}
\phantomsection\label{fseq.reporting:fseq.reporting.reports.ReportBase.saveFig}\pysiglinewithargsret{\bfcode{saveFig}}{\emph{fig}, \emph{outputRoot}, \emph{outputNamePrefix}, \emph{name=None}, \emph{*args}, \emph{**kwargs}}{}
Saves a figure and creates directories if needed.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\textbf{fig: matplotlib.figure}
\begin{quote}

The figure to be saved
\end{quote}

\textbf{outputRoot: str}
\begin{quote}

The root directory for reports
\end{quote}

\textbf{outputNamePrefix: str}
\begin{quote}

If the name of the file should be prepended by some string.
Normally added by the \emph{Report Builder}
\end{quote}

\textbf{name: str, optional}
\begin{quote}

A specific name for this figure-file.
If none supplied the default name for the report will be used,

(Default: Use the \code{self.name} of the instance)

\textbf{Note:} If none supplied and none set for instance,
\code{ValueError} is raised.
\end{quote}

\textbf{*args:}
\begin{quote}

Any arguments to be sent to \code{matplotlib.Figure.save}.
If none added the \code{ReportBase.saveArgs} will be used.
\end{quote}

\textbf{**kwargs:}
\begin{quote}

Any keyword arguments to be sent to \code{matplotlib.Figure.save}.
If notn addd the \code{ReportBase.saveKwargs} will be used.
\end{quote}

\item[{Returns}] \leavevmode
ReportBase
\begin{quote}

Returns \code{self}
\end{quote}

\item[{Raises}] \leavevmode
\textbf{ValueError}
\begin{quote}

If no name has been given.
\end{quote}

\end{description}\end{quote}

\end{fulllineitems}

\index{saveKwargs (fseq.reporting.reports.ReportBase attribute)}

\begin{fulllineitems}
\phantomsection\label{fseq.reporting:fseq.reporting.reports.ReportBase.saveKwargs}\pysigline{\bfcode{saveKwargs}}
Save keyword args passed to \code{matplotlib.pyplot.figure.savefig}:
dict

\end{fulllineitems}


\end{fulllineitems}



\subparagraph{Module contents}
\label{fseq.reporting:module-fseq.reporting}\label{fseq.reporting:module-contents}\index{fseq.reporting (module)}
Reporting-related modules of fseq.

The sub-package contains \emph{report\_builder} which post-processes data and
sends it off to make various \emph{reports}.

The builders contains no graphics information, but simply prepares data.
A new builder should be written if a new type of analysis is needed based
on an abstraction of the data that is not already present in any of the
previous builders.

A report makes an image from data sent to it from the builder.
Purpose of breaking this out is to allow for changing library that produces
the reports and to allow for quick reuse with similar graphics for identical
types of graphs for several report-builders.


\subsubsection{Module contents}
\label{fseq:module-contents}\label{fseq:module-fseq}\index{fseq (module)}
fSeq is a toolbox for sequence analysis in the frequency domain.

The module is organized around two phases of work: \emph{reading} and \emph{reporting}.

The reading uses a data format detector \emph{SeqFormatDetector}, which detects
the current \emph{SeqFormat} that is being read by the \emph{SeqReader} and passed to the
\emph{SeqEncoder} to translate into a numpy array

The \emph{Report{}`s and {}`ReportBuilder} use the output of the import phase to produce
the output. They are organized as such that a builder pre-processes the data
without any graphics done and the reporters the takes the pre-processed data
and make displays out of them.

All relevant parts of \emph{reading} and \emph{reporting} are directly imported to the
package root.


\paragraph{Reading}
\label{fseq:reading}
There is one generic reader that is intended to handle all use cases.
\begin{description}
\item[{fseq.SeqReader}] \leavevmode
Root object that coordinates reading, encoding and reporting

\end{description}

The encoders translates and manages format detection
\begin{description}
\item[{fseq.SeqEncoder}] \leavevmode
Base class encoder

\item[{fseq.SeqEncoderGC}] \leavevmode
Encoder that translates Gs and Cs to 1 while A and T become 0

\end{description}

There's a general format detector, and several data-formats.
\begin{description}
\item[{fseq.SeqFormatDetector}] \leavevmode
Detector of formats

\item[{fseq.SeqFormat}] \leavevmode
The base class from which all formats must be derived

\item[{fseq.FastaMultiline}] \leavevmode
Format detecting a fasta-file where sequence may span more than one line

\item[{fseq.FastaSingleline}] \leavevmode
Format detecting a fasta-file where sequence is in a signle line

\item[{fseq.FastQ}] \leavevmode
Format detecting fastq, but not which quality encoding

\end{description}


\paragraph{Reporting}
\label{fseq:reporting}
There is a report builder base from which all report builders should be
made, and two specific report builders.
\begin{description}
\item[{fseq.ReportBuilderBase}] \leavevmode
The base class for all builders

\item[{fseq.ReportBuilderPositionAverage}] \leavevmode
A report builder that averages data per position

\item[{fseq.ReportBuilderFFT}] \leavevmode
A report builder that subsamples and then does clustered FFT analysis

\end{description}

There are two reports included and an optional base class.
\begin{description}
\item[{fseq.ReportBase}] \leavevmode
Base class to make constructing new reports more efficient

\item[{fseq.LinePlot}] \leavevmode
Plots a line from the data sent to it

\item[{fseq.HeatMap}] \leavevmode
Plots a heat-map from the data sent to it.

\end{description}


\paragraph{Exceptions}
\label{fseq:exceptions}\begin{description}
\item[{fseq.FormatError}] \leavevmode
Base exception for error with data-formats

\item[{fseq.FormatImplementationError}] \leavevmode
If a format is not correctly implemented

\item[{fseq.FormatUnknown}] \leavevmode
If data is of unknown format

\end{description}


\section{Tutorial}
\label{tutorial::doc}\label{tutorial:tutorial}

\subsection{Installing}
\label{tutorial:installing}
The program is installed for current user by running:

\begin{Verbatim}[commandchars=\\\{\}]
\PYGZdl{} python setup.py install \PYGZhy{}\PYGZhy{}user
\end{Verbatim}

Or for all users by running:

\begin{Verbatim}[commandchars=\\\{\}]
\PYGZdl{} sudo python setup.py install
\end{Verbatim}

The following dependencies needs to be installed separately:

\begin{Verbatim}[commandchars=\\\{\}]
\PYG{n}{numpy}\PYG{p}{,} \PYG{n}{scipy}\PYG{p}{,} \PYG{n}{matplotlib}
\end{Verbatim}

On Debian systems copy:

\begin{Verbatim}[commandchars=\\\{\}]
\PYGZdl{} sudo apt\PYGZhy{}get update \PYGZam{}\PYGZam{} sudo apt\PYGZhy{}get install python\PYGZhy{}numpy python\PYGZhy{}scipy python\PYGZhy{}matplotlib
\end{Verbatim}


\subsection{Command Line Use}
\label{tutorial:command-line-use}
The following example runs default analysis on two different files:

\begin{Verbatim}[commandchars=\\\{\}]
\PYGZdl{} fseq \PYGZti{}/Data/Mysc\PYGZus{}24\PYGZus{}ATCACG\PYGZus{}L008\PYGZus{}R1\PYGZus{}001.fastq \PYGZti{}/Data/Mysc\PYGZus{}74\PYGZus{}GTTTCG\PYGZus{}L008\PYGZus{}R1\PYGZus{}001.fastq
14\PYGZhy{}06\PYGZhy{}23 17:55 SeqReader    INFO     Has 2 jobs
14\PYGZhy{}06\PYGZhy{}23 17:55 SeqReader    INFO     Reading: /home/martin/Data/Mysc\PYGZus{}24\PYGZus{}ATCACG\PYGZus{}L008\PYGZus{}R1\PYGZus{}001.fastq
14\PYGZhy{}06\PYGZhy{}23 17:57 SeqReader    INFO     Reading Complete: /home/martin/Data/Mysc\PYGZus{}24\PYGZus{}ATCACG\PYGZus{}L008\PYGZus{}R1\PYGZus{}001.fastq
14\PYGZhy{}06\PYGZhy{}23 17:57 SeqReader    INFO     Reporting \PYGZlt{}class \PYGZsq{}fseq.reporting.report\PYGZus{}builder.ReportBuilderFFT\PYGZsq{}\PYGZgt{} with args=(\PYGZdq{}\PYGZlt{}type \PYGZsq{}numpy.ndarray\PYGZsq{}\PYGZgt{}.shape=(4831521, 101)\PYGZdq{},), kwargs=\PYGZob{}\PYGZsq{}outputRoot\PYGZsq{}: \PYGZsq{}/home/martin/Data/Mysc\PYGZus{}24\PYGZus{}ATCACG\PYGZus{}L008\PYGZus{}R1\PYGZus{}001.fastq.reports\PYGZsq{}\PYGZcb{}
14\PYGZhy{}06\PYGZhy{}23 17:57 SeqReader    INFO     Reporting \PYGZlt{}class \PYGZsq{}fseq.reporting.report\PYGZus{}builder.ReportBuilderPositionAverage\PYGZsq{}\PYGZgt{} with args=(\PYGZdq{}\PYGZlt{}type \PYGZsq{}numpy.ndarray\PYGZsq{}\PYGZgt{}.shape=(4831521, 101)\PYGZdq{},), kwargs=\PYGZob{}\PYGZsq{}outputRoot\PYGZsq{}: \PYGZsq{}/home/martin/Data/Mysc\PYGZus{}24\PYGZus{}ATCACG\PYGZus{}L008\PYGZus{}R1\PYGZus{}001.fastq.reports\PYGZsq{}\PYGZcb{}
14\PYGZhy{}06\PYGZhy{}23 17:57 SeqReader    INFO     Reading: /home/martin/Data/Mysc\PYGZus{}74\PYGZus{}GTTTCG\PYGZus{}L008\PYGZus{}R1\PYGZus{}001.fastq
Saving \PYGZhy{}\PYGZgt{} /home/martin/Data/Mysc\PYGZus{}24\PYGZus{}ATCACG\PYGZus{}L008\PYGZus{}R1\PYGZus{}001.fastq.reports/fft\PYGZhy{}sample.abs.heatmap.pdf
Saving \PYGZhy{}\PYGZgt{} /home/martin/Data/Mysc\PYGZus{}24\PYGZus{}ATCACG\PYGZus{}L008\PYGZus{}R1\PYGZus{}001.fastq.reports/average.total.line.pdf
Saving \PYGZhy{}\PYGZgt{} /home/martin/Data/Mysc\PYGZus{}24\PYGZus{}ATCACG\PYGZus{}L008\PYGZus{}R1\PYGZus{}001.fastq.reports/fft\PYGZhy{}sample.angle.heatmap.pdf
14\PYGZhy{}06\PYGZhy{}23 18:01 SeqReader    INFO     Reading Complete: /home/martin/Data/Mysc\PYGZus{}74\PYGZus{}GTTTCG\PYGZus{}L008\PYGZus{}R1\PYGZus{}001.fastq
14\PYGZhy{}06\PYGZhy{}23 18:01 SeqReader    INFO     Reporting \PYGZlt{}class \PYGZsq{}fseq.reporting.report\PYGZus{}builder.ReportBuilderFFT\PYGZsq{}\PYGZgt{} with args=(\PYGZdq{}\PYGZlt{}type \PYGZsq{}numpy.ndarray\PYGZsq{}\PYGZgt{}.shape=(5634723, 101)\PYGZdq{},), kwargs=\PYGZob{}\PYGZsq{}outputRoot\PYGZsq{}: \PYGZsq{}/home/martin/Data/Mysc\PYGZus{}74\PYGZus{}GTTTCG\PYGZus{}L008\PYGZus{}R1\PYGZus{}001.fastq.reports\PYGZsq{}\PYGZcb{}
14\PYGZhy{}06\PYGZhy{}23 18:01 SeqReader    INFO     Reporting \PYGZlt{}class \PYGZsq{}fseq.reporting.report\PYGZus{}builder.ReportBuilderPositionAverage\PYGZsq{}\PYGZgt{} with args=(\PYGZdq{}\PYGZlt{}type \PYGZsq{}numpy.ndarray\PYGZsq{}\PYGZgt{}.shape=(5634723, 101)\PYGZdq{},), kwargs=\PYGZob{}\PYGZsq{}outputRoot\PYGZsq{}: \PYGZsq{}/home/martin/Data/Mysc\PYGZus{}74\PYGZus{}GTTTCG\PYGZus{}L008\PYGZus{}R1\PYGZus{}001.fastq.reports\PYGZsq{}\PYGZcb{}
14\PYGZhy{}06\PYGZhy{}23 18:01 SeqReader    INFO     Waiting for 4 report builders to finish
Saving \PYGZhy{}\PYGZgt{} /home/martin/Data/Mysc\PYGZus{}74\PYGZus{}GTTTCG\PYGZus{}L008\PYGZus{}R1\PYGZus{}001.fastq.reports/average.total.line.pdf
Saving \PYGZhy{}\PYGZgt{} /home/martin/Data/Mysc\PYGZus{}74\PYGZus{}GTTTCG\PYGZus{}L008\PYGZus{}R1\PYGZus{}001.fastq.reports/fft\PYGZhy{}sample.abs.heatmap.pdf
Saving \PYGZhy{}\PYGZgt{} /home/martin/Data/Mysc\PYGZus{}74\PYGZus{}GTTTCG\PYGZus{}L008\PYGZus{}R1\PYGZus{}001.fastq.reports/fft\PYGZhy{}sample.angle.heatmap.pdf
Saving \PYGZhy{}\PYGZgt{} /home/martin/Data/Mysc\PYGZus{}24\PYGZus{}ATCACG\PYGZus{}L008\PYGZus{}R1\PYGZus{}001.fastq.reports/average.lacking.line.pdf
Saving \PYGZhy{}\PYGZgt{} /home/martin/Data/Mysc\PYGZus{}24\PYGZus{}ATCACG\PYGZus{}L008\PYGZus{}R1\PYGZus{}001.fastq.reports/average.not\PYGZhy{}lacking.line.pdf
Saving \PYGZhy{}\PYGZgt{} /home/martin/Data/Mysc\PYGZus{}74\PYGZus{}GTTTCG\PYGZus{}L008\PYGZus{}R1\PYGZus{}001.fastq.reports/average.lacking.line.pdf
Saving \PYGZhy{}\PYGZgt{} /home/martin/Data/Mysc\PYGZus{}74\PYGZus{}GTTTCG\PYGZus{}L008\PYGZus{}R1\PYGZus{}001.fastq.reports/average.not\PYGZhy{}lacking.line.pdf
14\PYGZhy{}06\PYGZhy{}23 18:03 SeqReader    INFO     All jobs complete{}`
\end{Verbatim}

\textbf{Note:} Running above consumes quite a lot of memory and CPU and takes about
10 minutes.

\textbf{Note:} If \code{fseq} is not found on your system, it usually is due to the
default target of scripts for user install is not in your PATH.
To amend this, check where install copied the file \emph{scripts/fseq} and append
that to your current PATH.


\subsection{Python Use}
\label{tutorial:python-use}
For all scenarios it should suffice to import the package:

\begin{Verbatim}[commandchars=\\\{\}]
\PYG{g+gp}{\PYGZgt{}\PYGZgt{}\PYGZgt{} }\PYG{k+kn}{import} \PYG{n+nn}{fseq}
\end{Verbatim}

To create a reader that will run the analysis:

\begin{Verbatim}[commandchars=\\\{\}]
\PYG{g+gp}{\PYGZgt{}\PYGZgt{}\PYGZgt{} }\PYG{n}{r} \PYG{o}{=} \PYG{n}{fseq}\PYG{o}{.}\PYG{n}{SeqReader}\PYG{p}{(}\PYG{n}{dataSourcePaths}\PYG{o}{=}\PYG{p}{(}\PYG{l+s}{\PYGZdq{}}\PYG{l+s}{\PYGZti{}/Data/Mysc\PYGZus{}24\PYGZus{}ATCACG\PYGZus{}L008\PYGZus{}R1\PYGZus{}001.fastq}\PYG{l+s}{\PYGZdq{}}\PYG{p}{,} \PYG{l+s}{\PYGZdq{}}\PYG{l+s}{\PYGZti{}/Data/Mysc\PYGZus{}74\PYGZus{}GTTTCG\PYGZus{}L008\PYGZus{}R1\PYGZus{}001.fastq}\PYG{l+s}{\PYGZdq{}}\PYG{p}{)}\PYG{p}{)}
\end{Verbatim}

We can see how many jobs the reader has left:

\begin{Verbatim}[commandchars=\\\{\}]
\PYG{g+gp}{\PYGZgt{}\PYGZgt{}\PYGZgt{} }\PYG{n+nb}{len}\PYG{p}{(}\PYG{n}{r}\PYG{p}{)}
\PYG{g+go}{2}
\end{Verbatim}

And we can see the encoding that will be performed (and change it):

\begin{Verbatim}[commandchars=\\\{\}]
\PYG{g+gp}{\PYGZgt{}\PYGZgt{}\PYGZgt{} }\PYG{n}{r}\PYG{o}{.}\PYG{n}{seqEncoder}
\PYG{g+go}{\PYGZlt{}fseq.reading.seq\PYGZus{}encoder.SeqEncoderGC at 0x7fa9f9539b10\PYGZgt{}}
\end{Verbatim}

This encoder is the default and will translate Gs and Cs to 1 while As and Ts
are made into 0s.

We can also see which reports were requested by the encoder and thus added to
the reader since we didn't say what reports we wanted:

\begin{Verbatim}[commandchars=\\\{\}]
\PYG{g+gp}{\PYGZgt{}\PYGZgt{}\PYGZgt{} }\PYG{n+nb}{tuple}\PYG{p}{(}\PYG{n}{r}\PYG{o}{.}\PYG{n}{reportBuilders}\PYG{p}{)}
\PYG{g+go}{(\PYGZlt{}fseq.reporting.report\PYGZus{}builder.ReportBuilderFFT at 0x7fa9f95399d0\PYGZgt{},}
\PYG{g+go}{ \PYGZlt{}fseq.reporting.report\PYGZus{}builder.ReportBuilderPositionAverage at 0x7fa9f9539c50\PYGZgt{})}
\end{Verbatim}

To run encoding and produce results, simply:

\begin{Verbatim}[commandchars=\\\{\}]
\PYG{g+gp}{\PYGZgt{}\PYGZgt{}\PYGZgt{} }\PYG{n}{r}\PYG{o}{.}\PYG{n}{run}\PYG{p}{(}\PYG{p}{)}
\end{Verbatim}

Note that this will take some time and consume quite a lot of resources.
It took about 10 minutes on a standard desktop for the two files in the
command line example, and the python use is no different.


\section{Developers}
\label{developers:developers}\label{developers::doc}
\code{fseq} has been written to initially take care of a very limited set of
analysis and sequence formats, while at the same time be written to be
highly extensible.

Some examples of suitable features to be included in the future:
\begin{itemize}
\item {} 
\textbf{FastQ SeqFormat Subclasses}

Sub-classing {\hyperref[fseq.reading:fseq.reading.seq_encoder.FastQ]{\code{fseq.reading.seq\_encoder.FastQ}}} to automatically
detect which quality encoding
was used based on the range of values in the quality lines fed to it.

\item {} 
\textbf{SeqEncoderQaulity SeqEncoder Subclass}

Subclass {\hyperref[fseq.reading:fseq.reading.seq_encoder.SeqEncoder]{\code{fseq.reading.seq\_encoder.SeqEncoder}}} such that it
encodes the quality line using the quality encoding
supplied by the {\hyperref[fseq.reading:fseq.reading.seq_encoder.SeqFormat]{\code{fseq.reading.seq\_encoder.SeqFormat}}} detected.

\end{itemize}


\subsection{Git}
\label{developers:git}
The source code project is hosted at:

\href{https://gitorious.org/fseq}{https://gitorious.org/fseq}

For merge requests, the code is expected to:
\begin{itemize}
\item {} 
Be documented in accordance with \code{numpydoc}-format
(see \href{https://github.com/numpy/numpy/blob/master/doc/HOWTO\_DOCUMENT.rst.txt}{https://github.com/numpy/numpy/blob/master/doc/HOWTO\_DOCUMENT.rst.txt})

\item {} 
Code to be PEP8 compliant
(see \href{http://legacy.python.org/dev/peps/pep-0008/}{http://legacy.python.org/dev/peps/pep-0008/})

\item {} 
Coding style in general to follow the style established in \emph{fseq}

\item {} 
Git commits to in general be \emph{informative} and have
\emph{one aspect changed per commit}

\item {} 
Unit-tests to cover new functionality

\end{itemize}


\subsection{Reading}
\label{developers:reading}
To extend the functionality by adding further encoders, these encoders should
be derived from {\hyperref[fseq.reading:fseq.reading.seq_encoder.SeqEncoder]{\code{fseq.reading.seq\_encoder.SeqEncoder}}} and as a minimal
requirement need to overwrite the
{\hyperref[fseq.reading:fseq.reading.seq_encoder.SeqEncoder.parse]{\code{fseq.reading.seq\_encoder.SeqEncoder.parse()}}}-method.

To extend the functionality by adding support for more input formats, classes
should be derived from
{\hyperref[fseq.reading:fseq.reading.seq_encoder.SeqFormat]{\code{fseq.reading.seq\_encoder.SeqFormat}}} or any of the already implemented
formats if they partially solve detection for the new format.
Minimal requirement is overwriting the
{\hyperref[fseq.reading:fseq.reading.seq_encoder.SeqFormat.expects]{\code{fseq.reading.seq\_encoder.SeqFormat.expects()}}}-method, but
typically many of the properties of the base class as well as the
constructor needs replacing.


\subsubsection{\texttt{SeqEncoder.parse(self, lines, out, outindex)} overwriting}
\label{developers:seqencoder-parse-self-lines-out-outindex-overwriting}
\textbf{Important 1:} The overwritten
{\hyperref[fseq.reading:fseq.reading.seq_encoder.SeqEncoder.parse]{\code{fseq.reading.seq\_encoder.SeqEncoder.parse()}}} must have identical
parameter set. If further information is needed, this should be dealt with
during initiation or by separate methods.

\textbf{Important 2:} The overwritten method may not throw any errors and should silently
handle scenarios where the length of the information to be encoded mismatches
the length of the corresponding slot of the \code{out} object.

Example of how out the second important note can be achieved (adapted from
{\hyperref[fseq.reading:fseq.reading.seq_encoder.SeqEncoderGC.parse]{\code{fseq.reading.seq\_encoder.SeqEncoderGC.parse()}}}):

\begin{Verbatim}[commandchars=\\\{\}]
\PYG{c}{\PYGZsh{}Point to line of interes}
\PYG{n}{l} \PYG{o}{=} \PYG{n}{lines}\PYG{p}{[}\PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{\PYGZus{}sequence\PYGZus{}line}\PYG{p}{]}

\PYG{c}{\PYGZsh{}Put the contents of that line directly into out}
\PYG{n}{out}\PYG{p}{[}\PYG{n}{outindex}\PYG{p}{]}\PYG{p}{[}\PYG{p}{:}\PYG{n+nb}{len}\PYG{p}{(}\PYG{n}{l}\PYG{p}{)}\PYG{p}{]} \PYG{o}{=} \PYG{n}{l}\PYG{p}{[}\PYG{p}{:}\PYG{n}{out}\PYG{o}{.}\PYG{n}{shape}\PYG{p}{(}\PYG{l+m+mi}{1}\PYG{p}{)}\PYG{p}{]}
\end{Verbatim}

The above example is quite useless as an encoder as it doesn't translate
the contents of the input in any way, but the \code{{[}:len(l){]}} on \code{out}
ensures the target slot of \code{out} is not too large, while the
\code{{[}:out.shape(1){]}} ensures that \emph{l} is not too large for the slot in \code{out}.

\textbf{Important 3:} The \code{parse}-method may use the state of the class instance
(as in the above example), but due to concurrency issues, \emph{it should not alter
the state}.


\subsubsection{\texttt{SeqFormat} sub-classing}
\label{developers:seqformat-sub-classing}
** Important 1:** If a class is parent to further sub-classing such that the
class will conform to all data that the more specific subclass will do
(e.g. FastQ will be expect all lines/return \code{True} for all scenarios that
a FastQ\_Q33-subclass that detects fastq-files with encoding starting at 33),
then:
\begin{itemize}
\item {} 
The \emph{parent} should implement the
\code{fseq.reading.seq\_encoder.SeqFormat.\_decay()} method similar
to the base class and have a suitable \code{self.\_giveup} set in its init.

\item {} 
The specific \emph{child} should overwrite the \code{\_decay}-method so that it
never gives up \emph{or alternatively} takes longer before it gives up by
having ha higher number set to \code{self.\_giveup} in init.

\end{itemize}

The \code{self.expect(line)} should return a boolean if the line fits what was
expected as the next line, this method doesn't need to continue
reporting \code{False} after its first occurrence.
As soon as an \code{expect}-method returns a \code{False}, that \code{SeqFormat} is
removed from possible formats by the
{\hyperref[fseq.reading:fseq.reading.seq_encoder.SeqFormatDetector]{\code{fseq.reading.seq\_encoder.SeqFormatDetector}}}.


\subsection{Reporting}
\label{developers:reporting}
To extend the available abstractions/analysis done to the encoded
data, new derived {\hyperref[fseq.reporting:fseq.reporting.report_builder.ReportBuilderBase]{\code{fseq.reporting.report\_builder.ReportBuilderBase}}}
classes should be made.
Typically the \code{\_\_init\_\_} and \code{distill} would be overwritten (but the super
class methods called), and the \code{DEFAULT\_REPORTS} attribute replaced.
Potentially the interface extended by more relevant methods and properties
needed for user customization of the post-processing.

For creating new reports any object having a \code{distill}-method will do, but
using {\hyperref[fseq.reporting:fseq.reporting.reports.ReportBase]{\code{fseq.reporting.reports.ReportBase}}} will save some implementation
by having implemented the common aspects of saving figures in \code{matplotlib}.


\subsubsection{\texttt{ReportBuilderBase} sub-classing}
\label{developers:reportbuilderbase-sub-classing}
To maintain the constructor interface it is highly recommended that the init
has the following structure:

\begin{Verbatim}[commandchars=\\\{\}]
\PYG{k}{def} \PYG{n+nf}{\PYGZus{}\PYGZus{}init\PYGZus{}\PYGZus{}}\PYG{p}{(}\PYG{n+nb+bp}{self}\PYG{p}{,} \PYG{o}{*}\PYG{n}{reports}\PYG{p}{,} \PYG{o}{*}\PYG{o}{*}\PYG{n}{kwargs}\PYG{p}{)}\PYG{p}{:}

    \PYG{k}{if} \PYG{n+nb}{len}\PYG{p}{(}\PYG{n}{reports}\PYG{p}{)} \PYG{o}{==} \PYG{l+m+mi}{0}\PYG{p}{:}
        \PYG{n}{reports} \PYG{o}{=} \PYG{n+nb}{tuple}\PYG{p}{(}\PYG{n}{r}\PYG{p}{(}\PYG{p}{)} \PYG{k}{for} \PYG{n}{r} \PYG{o+ow}{in} \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{DEFAULT\PYGZus{}REPORTS}\PYG{p}{)}

    \PYG{n+nb}{super}\PYG{p}{(}\PYG{n}{MyReportBuilder}\PYG{p}{,} \PYG{n+nb+bp}{self}\PYG{p}{)}\PYG{o}{.}\PYG{n}{\PYGZus{}\PYGZus{}init\PYGZus{}\PYGZus{}}\PYG{p}{(}\PYG{o}{*}\PYG{n}{reports}\PYG{p}{,} \PYG{o}{*}\PYG{o}{*}\PYG{n}{kwargs}\PYG{p}{)}

    \PYG{c}{\PYGZsh{}Emulating default values for keywords is done by getting}
    \PYG{c}{\PYGZsh{}the key with default values as follows}
    \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{someKey} \PYG{o}{=} \PYG{n}{kwargs}\PYG{o}{.}\PYG{n}{get}\PYG{p}{(}\PYG{l+s}{\PYGZsq{}}\PYG{l+s}{someKey}\PYG{l+s}{\PYGZsq{}}\PYG{p}{,} \PYG{n}{defaultValue}\PYG{p}{)}
\end{Verbatim}

To push some data to all attached reports make a \code{super} call to
{\hyperref[fseq.reporting:fseq.reporting.report_builder.ReportBuilderBase.distill]{\code{fseq.reporting.report\_builder.ReportBuilderBase.distill()}}}.


\subsubsection{\texttt{ReportBase} sub-classing or not}
\label{developers:reportbase-sub-classing-or-not}
Using \code{ReportBase} to create new reports is entirely optional, but if the
report is a \code{matplotlib}-report, then it is probably useful.

If sub-classing, then the
{\hyperref[fseq.reporting:fseq.reporting.reports.ReportBase.distill]{\code{fseq.reporting.reports.ReportBase.distill()}}} must be overwritten and
subclass should use a call to the inherited \code{saveFig}-method to do the
actually saving once the figure has been setup within the \code{distill} method.

If using other modules than \code{matplotlib} and thereby not sub-classing
\code{ReportBase}, the report should as a minimal requirement have a
\code{distill} method that takes the main data as the first argument and that
accepts any number of argument and keyword arguments by having something like
\code{*args, **kwargs} at the end of the parameter list.


\section{fSeq Project Report for C3SE Graduate Course: Python and High Performance Computing 2014}
\label{c3se_python_course::doc}\label{c3se_python_course:fseq-project-report-for-c3se-graduate-course-python-and-high-performance-computing-2014}

\subsection{Abstract}
\label{c3se_python_course:abstract}
\emph{fSeq} was created for the C3SE Graduate Course {\hyperref[c3se_python_course:c3se]{{[}c3se{]}}}, in part for want of
suitable existing project or code base.
Therefor only part of the course content
could be covered -- \code{numpy}, \code{scipy}, \code{matplotlib}, \code{unittest} and
\code{sphinx} being the most prominent.
The package contains classes to do simple per base analysis of the data as is
existing elsewhere in e.g. {\hyperref[c3se_python_course:fastqc]{{[}fastqc{]}}}.
However, it also extends sequence quality analysis with heat maps of
clustered Fourier data, which to the best of my knowledge, is novel to the
field.
These reflect previously identified issues of sequence data but also introduces
new frequency features, which requires further investigation.


\subsection{Solution Method}
\label{c3se_python_course:solution-method}

\subsubsection{Design Analysis}
\label{c3se_python_course:design-analysis}
The task was split up into several well defined components:
\begin{itemize}
\item {} 
\textbf{Reading data}

This feature can be held generic, it will need to be able to talk to
both encoders and format detectors on one side as well as post-processing
classes on the other.

The class can do several parts in parallel:
\begin{itemize}
\item {} 
Reading data from file

\item {} 
Encoding chunks of data

\item {} 
Reporting on encoded data

\end{itemize}

\item {} 
\textbf{Detecting input data format}

To be able to support several formats and allow for future extension of
formats supported, the task should be separated into \emph{selecting suitable
detector} among a collection of format collectors and \emph{detecting particular
formats}.

The latter should have a common interface for the former to use, thus a
base class that the detectors can subclass is suitable.

The logic of the formats differ enough that making a factory design pattern
would probably require more than is gained.

\item {} 
\textbf{Encoding input}

Potentially there is an unknown number of ways to encode the data.

The data reader needs an interface where objects can be sent such that
no concurrency issues can arise, that is, it may not alter the state.
This interface needs to be common to all encoders.

To simplify the use, the encoder can manage its format detector

The encoder can suggest default post-processors to be coupled with it in
the reader.
This will make the invocation less transparent, but greatly reduce the
workload of the user.
Thus the user must be able to override the default coupling, and this
manner must be clear.

\item {} 
\textbf{Post-processing encoded input}

One type of encoding can potentially be used for several downstream analyses.
However, it is not inherently clear if some aspects should be done by the
encoder directly or the post-processing class.

The post-processing class should coordinate all outputs made from its data.

It should assist in naming and annotation to clarify the contents of the
report.

Reports from one post-processor should be grouped by file name.

Post-processors need to have a common interface that doesn't alter the state
for the reader to use.

\item {} 
\textbf{Producing outputs}

Dependence on graphics libraries such as \code{matplotlib} should be restricted
to the output generators.

These classes also need to have a common interface for the post-processors
and may not alter the state (due to concurrencies).

An output producer should be reusable for several post-processors.

\end{itemize}


\paragraph{General}
\label{c3se_python_course:general}
The default invocation of analysis should require a minimum of user input.
For each deviation for what is considered \emph{default}, a new set of default
behaviors should exist.

The interdependence of the classes should be kept simple and clean such that
each class can be instantiated and as a maximum be needed as a parameter to
one other class.

The complete setup and configuration of a class should be possible via its
constructor and methods returning \code{self} such that all imaginable
combinations of settings can be setup and run in a single line.


\subsubsection{Parallelism}
\label{c3se_python_course:parallelism}
As described above, the main target for explicit coding of parallel computing
is the reader object.

To be able to easily share the data storing \code{numpy.ndarray}, \code{treading} was
selected. Other considered modules were \code{multiprocessing} and \code{pycuda},
however due to limitations in implementation time the former was used.

Still parallelism poses several issues, the size of the array cannot be
changed without creating a new object, while the needed size cannot be known
before the contents of the file has been read.
If the whole file is scanned, reading is impossible due to the large amount of
memory needed, a large benefit of threading would have been lost.

Therefore, a design was opted for where the main thread creates a large array
and several threads are used to fill that array up from data read from the
source file.
When the array is full, the main thread must pause, reading data and wait for
all live threads to finish, after which the array can be extended.
Then, reading and threading can start again.


\subsubsection{Unittesting}
\label{c3se_python_course:unittesting}
Test driven development {\hyperref[c3se_python_course:tdd]{{[}tdd{]}}} was considered, but as development and
especially design time was extremely limited, testing was decided to:
\begin{itemize}
\item {} 
Verify all isolated behaviors

\item {} 
Verify all interdependent behaviors that don't require running the
entire analysis nor would produce files on the hard drive.

\end{itemize}

The tests were decided to be placed inside the package but not be part
of the distribution.


\subsubsection{Documentation}
\label{c3se_python_course:documentation}
The documentation of the code should be compatible with \code{sphinx} {\hyperref[c3se_python_course:sphinx]{{[}sphinx{]}}}
and follow the \code{numpydoc} {\hyperref[c3se_python_course:npd]{{[}npd{]}}} standard of restructured text {\hyperref[c3se_python_course:rst]{{[}rst{]}}}.


\subsection{Implementation}
\label{c3se_python_course:implementation}

\subsubsection{Package structure}
\label{c3se_python_course:package-structure}
The relevant folder tree for the package was devised as follows:
\begin{itemize}
\item {} 
fseq (root of \emph{git}-repository)
\begin{itemize}
\item {} 
fseq (package/source root)
\begin{itemize}
\item {} 
reading

\item {} 
reporting

\item {} 
tests (testings modules, not included in distribution)

\end{itemize}

\item {} 
scripts (run-scripts installed)

\item {} 
doc (sphinx-documentation)

\end{itemize}

\end{itemize}

The \emph{setup.py} file was structured so that the scripts in the script folder
were installed as executables so that the package can be run as a stand
alone command line program.

A \emph{MANIFEST.in} was created in accordance with \code{distutil}`s recommendations
{\hyperref[c3se_python_course:distutil]{{[}distutil{]}}} to allow for distribution of packages via the \emph{setup.py} file.
The tests in the \emph{testing} folder were purposely kept out of packaging as they
were not considered part of the deployment code, but rather the development
source code.


\subsubsection{Design}
\label{c3se_python_course:design}
The structure and interfaces of the classes kept as designed, making the
following basic types:
\begin{itemize}
\item {} 
\code{SeqReader}

\item {} 
\code{SeqEncoder} to encode data and manage format detection if not predefined.

A specific subclass \code{SeqEncoderGC} was made to fulfill the goal of doing
GC-analysis

\item {} 
\code{SeqFormat} the object that detects specific formats for which three
different formats are supported \code{FastaSingleline}, \code{FastaMultiline},
and \code{FastQ}

\item {} 
\code{SeqFormatDetector} to select which format an input stream is.

\item {} 
\code{ReportBuilderBase} the post-processing coordinator, for which two
specific post-processors were created to allow \code{fseq} to produce usable
Fourier reports: \code{ReportBuilderFFT} and \code{ReportBuilderPositionAverage}.

\item {} 
\code{ReportBase} conforms with output producer, for which two specific
graph producers (\code{LinePlot} and \code{HeatMap}) were created.

\end{itemize}

To comply with the general design criteria, all relevant classes are imported
into the package root such that the user only needs to use \code{import fseq}.

Default behavior is simple as the following is sufficient:

\begin{Verbatim}[commandchars=\\\{\}]
\PYG{g+gp}{\PYGZgt{}\PYGZgt{}\PYGZgt{} }\PYG{n}{fseq}\PYG{o}{.}\PYG{n}{SeqReader}\PYG{p}{(}\PYG{n}{dataSourcePaths}\PYG{o}{=}\PYG{l+s}{\PYGZdq{}}\PYG{l+s}{some/path/to/file.fastq}\PYG{l+s}{\PYGZdq{}}\PYG{p}{)}\PYG{o}{.}\PYG{n}{run}\PYG{p}{(}\PYG{p}{)}
\end{Verbatim}

Further, full customization can be performed and expressed in a single line.
The expression can also be split to several lines increase readability.


\subsubsection{Unittests}
\label{c3se_python_course:unittests}
In total 78 different tests were created in four different files.
Each file corresponding to one of the four modules in the package.
A test exclusively tested one aspect of the functionality, but many of the tests
asserted more than one behavior for that aspect.

For example, \code{TestSeqFormatDetector.test\_FormatUnknown} that ascertains that
an exception is raised for when the detector runs out of available formats both
when it was initiated with and without a forced format.


\subsubsection{Documentation}
\label{c3se_python_course:id8}
All classes were fully documented as decided and several \code{sphinx} used to
produce a complete documentation with several supporting extra documents.


\subsection{Results}
\label{c3se_python_course:results}

\subsubsection{Technical results}
\label{c3se_python_course:technical-results}
A run took less than 10 minutes on a standard Intel i5 desktop with 4GB
RAM and a 2TB HDD. Typically more than 100\% CPU was used, though during
resizing of the array, a dipping of CPU was clear due to main thread waiting
for all threads to join. The memory usage peaked around 75\% when using 16-bit
float point precision, in \emph{numpy}.
With default settings, five report pdf:s were created for each file analyzed.

The unit tests typically ran for a fraction of a second and succeeded in reporting
previously undetected errors as well as alerting to inconsistencies caused by
minor changes of interfaces during development.


\subsubsection{Analysis of two files}
\label{c3se_python_course:analysis-of-two-files}
Two real data files were analyzed \emph{Mysc\_24\_ATCACG\_L008\_R1\_001.fastq} and
\emph{Mysc\_74\_GTTTCG\_L008\_R1\_001.fastq}.
The two files were multiplexed in the same Illumina MiSeq lane, but are two
distinct species.
Therefore, technical aspects of the sequencing can possibly be seen as
recurring features in the two, while aspects pertaining to the DNA in each
sample should be private.

As an example, the occurrence of undecided nucleotides is highly concurrent in
both data files:
\code{Mysc 24}
\code{Mysc 74}

While the GC bias over the two files are distinctly different:
\code{Mysc 24}
\code{Mysc 74}

The \emph{Myst 24} having a highly structured bias as averaged over the \textasciitilde{}5M reads.

The random sample of 1000 reads, Fourier Transformed and clustered based on
their amplitudes show little obvious structure in their angles:

\code{Mysc 24}
\code{Mysc 74}

While the corresponding amplitudes for the same 1000 reads share two clear
features. First, for the 0-frequency, an obvious large spread in overall GC
bias is evident with a small subset of around 90\% GC a majority around 40-50
and another smaller cluster close to 0\%. The second feature, which shows clearly
in both is that the 1/34 frequency and its neighbors behave distinctively.

\code{Mysc 24}
\code{Mysc 74}


\subsection{Discussion}
\label{c3se_python_course:discussion}

\subsubsection{Package}
\label{c3se_python_course:package}
The general design of the project was maintained during development and the
extension of functionality during worked as intended.
The package therefore shows promise of being well structured and designed.

The \code{threading} had some inherent issues with sleeping threads not appearing
alive causing jumbled and random encodings initially until sufficiently slow
implementation ensured threads are truly joined before reshaping of encoding
array.
There are some possibilities for further improving the performance of the
\code{SeqReader} by decoupling the data reading from the managing of the encoding
threads as well as taking an active part in managing the number of the latter.
Moving away from single processing should also be feasible and could be the
target of further performance development.

The use of unit tests worked well in assisting the development and as they were
written in junction with the code they were not merely a \emph{post-hoc} addition to
prove the correctness of the implementation, but actively discovered issues
previously unknown.

In general, the time plan was kept with the exception of documentation and
report writing, for which much more time would have been needed to learn
\code{sphinx} and \code{numpydoc} sufficiently well to produce both this report
and the general package documentation.


\subsubsection{Bioinformatics}
\label{c3se_python_course:bioinformatics}
The analyses included in the package reproduces know result where comparison is
applicable.
For example, the uneven bias of GC initially due to faulty timing of adapters
-- a known issue.
More interestingly the implicated a recurring frequency on the amplitude
analysis of clustered FFT data around 34/101.
The implication of this needs to be further investigated.
Potentially, protein coding regions in the sequence, for which triplicates of
nucleotides form the information unit in translation of DNA to amino acids of
the protein, could be related as it implies the factor 3.
However, \emph{why} and if this information can be useful remains to be investigated.


\subsection{References}
\label{c3se_python_course:references}

\subsection{Appendix A: Project Plan}
\label{c3se_python_course:appendix-a-project-plan}
The \code{project plan} submitted for the project.


\subsection{Appendix B: Code}
\label{c3se_python_course:appendix-b-code}
The current code is accessible from \emph{Gitorious} at:

\href{https://gitorious.org/fseq}{https://gitorious.org/fseq}

Alternatively, each class implementation can be accessed here:
\begin{itemize}
\item {} 
{\hyperref[fseq.reading:module-fseq.reading]{\code{fseq.reading}}}
\begin{quote}

{\hyperref[fseq.reading:fseq.reading.seq_reader.SeqReader]{\code{fseq.reading.seq\_reader.SeqReader}}}

{\hyperref[fseq.reading:fseq.reading.seq_encoder.SeqEncoder]{\code{fseq.reading.seq\_encoder.SeqEncoder}}}

{\hyperref[fseq.reading:fseq.reading.seq_encoder.SeqFormat]{\code{fseq.reading.seq\_encoder.SeqFormat}}}
\begin{quote}

{\hyperref[fseq.reading:fseq.reading.seq_encoder.FastQ]{\code{fseq.reading.seq\_encoder.FastQ}}}

{\hyperref[fseq.reading:fseq.reading.seq_encoder.FastaMultiline]{\code{fseq.reading.seq\_encoder.FastaMultiline}}}

{\hyperref[fseq.reading:fseq.reading.seq_encoder.FastaSingleline]{\code{fseq.reading.seq\_encoder.FastaSingleline}}}
\end{quote}

{\hyperref[fseq.reading:fseq.reading.seq_encoder.SeqFormatDetector]{\code{fseq.reading.seq\_encoder.SeqFormatDetector}}}
\end{quote}

\item {} 
{\hyperref[fseq.reporting:module-fseq.reporting]{\code{fseq.reporting}}}
\begin{quote}

{\hyperref[fseq.reporting:fseq.reporting.reports.ReportBase]{\code{fseq.reporting.reports.ReportBase}}}
\begin{quote}

{\hyperref[fseq.reporting:fseq.reporting.reports.HeatMap]{\code{fseq.reporting.reports.HeatMap}}}

{\hyperref[fseq.reporting:fseq.reporting.reports.LinePlot]{\code{fseq.reporting.reports.LinePlot}}}
\end{quote}

{\hyperref[fseq.reporting:fseq.reporting.report_builder.ReportBuilderBase]{\code{fseq.reporting.report\_builder.ReportBuilderBase}}}
\begin{quote}

{\hyperref[fseq.reporting:fseq.reporting.report_builder.ReportBuilderFFT]{\code{fseq.reporting.report\_builder.ReportBuilderFFT}}}

{\hyperref[fseq.reporting:fseq.reporting.report_builder.ReportBuilderPositionAverage]{\code{fseq.reporting.report\_builder.ReportBuilderPositionAverage}}}
\end{quote}
\end{quote}

\end{itemize}


\section{License}
\label{license::doc}\label{license:license}
The MIT License (MIT)

Copyright (c) 2014 Martin Zackrisson

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the ``Software''), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED ``AS IS'', WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.


\chapter{Indices}
\label{index:indices}\begin{itemize}
\item {} 
\emph{genindex}

\item {} 
\emph{modindex}

\end{itemize}


\chapter{Contact}
\label{index:contact}
The package was written and is maintained by Martin Zackrisson.

Developers, look at {\hyperref[developers::doc]{\emph{Developers}}} about Git and use \emph{Gitorious} as
mode of contact.

For end-users, questions can be addressed to the e-mail:
\begin{quote}

martin{[}dot{]}zackrisson{[}at{]}gu{[}dot{]}se
\end{quote}

\begin{thebibliography}{distutil}
\bibitem[c3se]{c3se}{\phantomsection\label{c3se_python_course:c3se} 
\href{http://www.c3se.chalmers.se/index.php/Python\_and\_High\_Performance\_Computing\_2014}{http://www.c3se.chalmers.se/index.php/Python\_and\_High\_Performance\_Computing\_2014}
}
\bibitem[distutil]{distutil}{\phantomsection\label{c3se_python_course:distutil} 
\href{https://docs.python.org/2/distutils/sourcedist.html\#the-manifest-in-template}{https://docs.python.org/2/distutils/sourcedist.html\#the-manifest-in-template}
}
\bibitem[tdd]{tdd}{\phantomsection\label{c3se_python_course:tdd} 
\href{http://en.wikipedia.org/wiki/Test-driven\_development}{http://en.wikipedia.org/wiki/Test-driven\_development}
}
\bibitem[npd]{npd}{\phantomsection\label{c3se_python_course:npd} 
\href{https://github.com/numpy/numpy/blob/master/doc/HOWTO\_DOCUMENT.rst.txt\#common-rest-concepts}{https://github.com/numpy/numpy/blob/master/doc/HOWTO\_DOCUMENT.rst.txt\#common-rest-concepts}
}
\bibitem[sphinx]{sphinx}{\phantomsection\label{c3se_python_course:sphinx} 
\href{http://sphinx-doc.org/}{http://sphinx-doc.org/}
}
\bibitem[rst]{rst}{\phantomsection\label{c3se_python_course:rst} 
\href{http://docutils.sourceforge.net/docs/ref/rst/restructuredtext.html\#bullet-lists}{http://docutils.sourceforge.net/docs/ref/rst/restructuredtext.html\#bullet-lists}
}
\bibitem[fastqc]{fastqc}{\phantomsection\label{c3se_python_course:fastqc} 
\href{http://www.bioinformatics.babraham.ac.uk/projects/fastqc/}{http://www.bioinformatics.babraham.ac.uk/projects/fastqc/}
}
\end{thebibliography}


\renewcommand{\indexname}{Python Module Index}
\begin{theindex}
\def\bigletter#1{{\Large\sffamily#1}\nopagebreak\vspace{1mm}}
\bigletter{f}
\item {\texttt{fseq}}, \pageref{fseq:module-fseq}
\item {\texttt{fseq.reading}}, \pageref{fseq.reading:module-fseq.reading}
\item {\texttt{fseq.reading.seq\_encoder}}, \pageref{fseq.reading:module-fseq.reading.seq_encoder}
\item {\texttt{fseq.reading.seq\_reader}}, \pageref{fseq.reading:module-fseq.reading.seq_reader}
\item {\texttt{fseq.reporting}}, \pageref{fseq.reporting:module-fseq.reporting}
\item {\texttt{fseq.reporting.report\_builder}}, \pageref{fseq.reporting:module-fseq.reporting.report_builder}
\item {\texttt{fseq.reporting.reports}}, \pageref{fseq.reporting:module-fseq.reporting.reports}
\end{theindex}

\renewcommand{\indexname}{Index}
\printindex
\end{document}
